---
title: "IBS Analysis"
author: "Mike Connelly"
date: "2023-05-10"
output: html_document
---
Code for *Pocillopora* genome skimming manuscript

## Setup and packages
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, cache = TRUE)
options(stringsAsFactors = FALSE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```
```{r packages}
library("tidyverse")
library("magrittr")
library("ggtree")
#
library("WGCNA") # for coloring clonal groups
library("sparcl") # for ColorDendrogram
# 
source("./R/pocillopora_etp_gksim_functions.R")
```

## Import sample metadata
```{r import_sample_metadata}
source("./R/sample_import.R")
samples <- all_samples
```
```{r select_samples}
samples <- filter(all_samples, ANALYSIS ==T)
# samples <- filter(all_samples, `PANAMA_MANUSCRIPT` == T)
```

```{r bamfile_sample_order_check}
# reading list of bam files = order of samples in IBS matrix
bams <- read.table("/Users/mikeconnelly/computing/projects/etp_pocillopora_gskim/outputs/angsd/full_analysis_bamfile.txt",header=F)[,1]
#
bam_ids <- sub("/scratch/nmnh_corals/connellym/projects/etp_pocillopora_gskim/outputs/alignments/", "", bams, perl=T)
bam_ids <- sub(".sorted.md.rg.bam", "", bam_ids, perl=T)
bam_ids <- sub("_S[[:digit:]]+$", "", bam_ids)
#
bam_ids
```
```{r bam_sample_order_check}
# are all the bam files in the sample metadata?
all(bam_ids %in% samples$`Sample ID`) # TRUE
# which bam files are not in the sample metadata?
bam_ids[bam_ids %notin% samples$`Sample ID`] # character(0)
#
# are all samples in the bam files?
all(samples$`Sample ID` %in% bam_ids) # TRUE
# which samples are not in the bam files?
samples[samples$`Sample ID` %notin% bam_ids,]
#
# do all the sample ids match the bam file ids in the correct order?
all(samples$`Sample ID` == bam_ids) # TRUE
# which samples ids do not match? 
samples[samples$`Sample ID` != bam_ids,]
#
# aligning sample metadata with bams list, then check again
samples <- samples[match(bam_ids, samples$`Sample ID`),] 
# do all the sample ids match the bam file ids in the correct order?
all(samples$`Sample ID` == bam_ids) # TRUE
```
# All analysis samples, clones included
## Import IBS matrix
```{r import_ibs_matrix}
# reading IBS matrix based on SNPs with allele frequency >= 0.05:
# ma <- as.matrix(read.table("./outputs/angsd/etp_ibs05.ibsMat"))
# ma <- as.matrix(read.table("./outputs/angsd/panama_manuscript_ibs05.ibsMat"))
# ma <- as.matrix(read.table("./outputs/angsd/all_ibs05.ibsMat"))
# ma <- as.matrix(read.table("./outputs/angsd/round2_ibs05.ibsMat"))
# ma <- as.matrix(read.table("./outputs/angsd/all_analysis_ibs05.ibsMat"))
# ma <- as.matrix(read.table("./outputs/angsd/costarica_ibs05.ibsMat"))
ma <- as.matrix(read.table("./outputs/angsd/full_analysis_ibs05.ibsMat"))

```
```{r import_ibs_matrix}
# remove biorepository codes from sample labels
sample_names <- gsub("_AP3X.*", "", samples$`Sample ID`)
# IMPORTANT: verify that the order of ids in sample metadata and in the bam file list used for IBS in ANGSD match
dimnames(ma) <- list(samples$`Sample ID`, samples$`Sample ID`)
# dimnames(ma) <- list(sample_names, sample_names)
#
# dimnames(ma)=list(samples$Site,samples$Site)
# OR
# dimnames(ma) <- list(bam_ids, bam_ids)
```
### Plot IBS hierarchical clustering dendrograms
```{r ibs_dendrogram_test}
# plotting hierarchical clustering dendrogram
hc <- hclust(as.dist(ma), "ave")
plot(hc, cex = 0.4)
```
### IBS matrix subsets dendrograms
```{r ibs_dendrogram_rsmas}
# retaining only rsmas samples
ma_rsmas <- ma[samples$RSMAS,samples$RSMAS]
# ma_rsmas <- ma[samples$REPLICATE,samples$REPLICATE]
# re-clustering
hc_rsmas <- hclust(as.dist(ma_rsmas),"ave")
plot(hc_rsmas,cex=0.4)
abline(h=0.145,col="darkorange",lty=3)
abline(h=0.105,col="turquoise2",lty=3)
```
```{r ibs_dendrogram_stri}
# retaining only STRI samples - these have pictures, Panama have skeleton vouchers!
ma_stri <- ma[samples$STRI, samples$STRI]
# re-clustering
hc_stri <- hclust(as.dist(ma_stri),"ave")
plot(hc_stri, cex=0.4)
abline(h=0.145,col="darkorange",lty=3)
abline(h=0.105,col="turquoise2",lty=3)
```
```{r ibs_dendrogram_manuscript}
pdf("./outputs/figures/Fig2_ibs_dendrogram_species_basic.pdf", height = 6, width = 6.5)
# retaining only Panama samples (STRI, RSMAS, Ana's) and JFF Clipperton samples - these have pictures, Panama have skeleton vouchers, or are already published
ma_panpopgen <- ma[samples$PANAMA_MANUSCRIPT, samples$PANAMA_MANUSCRIPT]
# re-clustering
hc_panpopgen <- hclust(as.dist(ma_panpopgen),"ave")
plot(hc_panpopgen, cex=0.4)
abline(h=0.145,col="darkorange",lty=3)
abline(h=0.105,col="turquoise2",lty=3)
dev.off()
```
```{r ibs_dendrogram_galapagos}
# retaining only galapagos samples
ma_usfq <- ma[samples$Set == "USFQ_CSUMB", samples$Set == "USFQ_CSUMB"]
# re-clustering
hc_usfq <- hclust(as.dist(ma_usfq),"ave")
plot(hc_usfq,cex=0.4)
abline(h=0.145,col="darkorange",lty=3)
abline(h=0.105,col="turquoise2",lty=3)
```
```{r ibs_dendrogram_coco}
# retaining only isla del coco samples
ma_coco <- ma[samples$Region == "Coco",samples$Region == "Coco"]
# re-clustering
hc_coco <- hclust(as.dist(ma_coco),"ave")
plot(hc_coco,cex=0.6)
abline(h=0.15,col="red",lty=3)
```

## Plot IBS clustering dendrogram with ggtree for all samples
```{r}
tree_title <- expression(paste("Identity-by-state clustering of all ", italic("Pocillopora"), " samples (n=", 337, ")"))
```
```{r ibs_ggtree}
# pdf("./outputs/figures/ibs_dendrogram_all_analysis_site.pdf",width = 8.5, height = 16)
hc %>% ggtree() %<+% samples + 
  # geom_vline(xintercept = -0.15, stroke = 1, color = "red", linetype = 2) +
  geom_tippoint(aes(fill = `Region:Site`), shape = 21, alpha = 0.8, stroke = 0.5, color = "black", hjust = 0.3, show.legend = TRUE) +
  geom_tiplab(size = 2, hjust = -0.05) +
  ggplot2::xlim(-0.45, 0.01) +
  geom_treescale(x = -0.3, width = 0.05) +
  # scale_fill_manual(values = site_colors) +
   theme(legend.position = c(0.2, 0.85),
        legend.margin = margin(0,0,0,0, unit = "mm"),
        plot.margin = margin(5,5,5,5, unit = "mm")) +
  ggtitle(tree_title)
# dev.off()
```
```{r}
# pdf("./outputs/figures/ibs_dendrogram_all_analysis_species.pdf", width = 8.5, height = 5)
hc %>% ggtree() %<+% samples + 
  layout_dendrogram() + 
  # geom_vline(xintercept = -0.15, stroke = 0.5, color = mtorf_colors[1], linetype = 2) +
  # geom_vline(xintercept = -0.05, stroke = 0.5, color = mtorf_colors[2], linetype = 2) +
  # geom_vline(xintercept = -0.105, stroke = 0.5, color = mtorf_colors[3], linetype = 2) +
  # geom_vline(xintercept = -0.103, stroke = 0.5, color = mtorf_colors[4], linetype = 2) +
  geom_tippoint(aes(fill = `Species`), shape = 21, alpha = 0.8, stroke = 0.5, color = "black", hjust = 0.3, show.legend = TRUE) +
  # geom_tiplab(size = 2, hjust = 1.05) +
  scale_fill_manual(values = spp_colors, name = "Species") +
  theme_dendrogram(legend.position = c(0.8, 0.85),
        legend.margin = margin(0,0,0,0, unit = "mm"),
        plot.margin = margin(5,5,5,5, unit = "mm")) +
  ggtitle(expression(paste(italic("Pocillopora"), " identity-by-state clustering dendrogram")), subtitle = "All East Pacific samples (n= 337)")
# dev.off()
```

### No Bones IBS figures

```{r}
pdf("./outputs/figures/sicb/ibs_dendrogram_all_analysis_black.pdf", width = 8.5, height = 4.85)
hc_black <- hc %>% ggtree() %<+% samples + 
  layout_dendrogram() + 
  geom_tippoint(shape = 21, fill = "grey40", color = "black", alpha = 0.8, stroke = 0.5, hjust = 0.3, show.legend = FALSE) +
  # geom_tiplab(size = 2, hjust = 1.05) +
  scale_y_continuous(expand = expansion(mult = c(0.02, 0))) +
  theme_dendrogram(legend.position = c(0.75, 0.85),
        legend.margin = margin(0,0,0,0, unit = "mm"),
        plot.margin = margin(2,2,2,2, unit = "mm")) +
  xlab("Genetic distance (1-IBS)")
dev.off()
```
```{r}
png("./outputs/figures/sicb/ibs_dendrogram_all_analysis_black.png", width = 8.5, height = 4.85, units = "in", type = "cairo-png", res = 300)
hc_black
dev.off()
```

```{r}
pdf("./outputs/figures/sicb/ibs_dendrogram_all_analysis_mtorf.pdf", width = 8.5, height = 4.85)
hc_mtorf <- hc %>% ggtree() %<+% samples + 
  layout_dendrogram() + 
  geom_tippoint(aes(fill = `mtorf_type`), shape = 21, color = "black", alpha = 0.8, stroke = 0.5, hjust = 0.3, show.legend = TRUE) +
  # geom_tiplab(size = 2, hjust = 1.05) +
  scale_fill_manual(values = mtorf_colors, name = "mtORF haplotype") +
  scale_y_continuous(expand = expansion(mult = c(0.02, 0))) +
  theme_dendrogram(legend.position = c(0.9, 0.82),
        legend.margin = margin(0,0,0,0, unit = "mm"),
        plot.margin = margin(2,2,2,2, unit = "mm")) +
  xlab("Genetic distance (1-IBS)")
dev.off()
```
```{r}
png("./outputs/figures/sicb/ibs_dendrogram_all_analysis_mtorf.png", width = 8.5, height = 4.85, units = "in", type = "cairo-png", res = 300)
hc_mtorf
dev.off()
```

```{r}
pdf("./outputs/figures/sicb/ibs_dendrogram_all_analysis_mtorf_replicates.pdf", width = 8.5, height = 4.85)
hc_mtorf_rep <- hc %>% ggtree() %<+% samples + 
  layout_dendrogram() + 
  geom_tippoint(aes(fill = `mtorf_type`, color = `REPLICATE`), shape = 21, alpha = 0.8, stroke = 0.5, hjust = 0.3, show.legend = F) +
  scale_fill_manual(values = mtorf_colors, name = "mtORF haplotype") +
  scale_color_manual(values = c("black", "red")) +
  scale_y_continuous(expand = expansion(mult = c(0.02, 0))) +
  theme_dendrogram(legend.position = c(0.9, 0.82),
        legend.margin = margin(0,0,0,0, unit = "mm"),
        plot.margin = margin(2,2,2,2, unit = "mm")) +
  xlab("Genetic distance (1-IBS)")
dev.off()
```
```{r}
png("./outputs/figures/sicb/ibs_dendrogram_all_analysis_mtorf_reps.png", width = 8.5, height = 4.85, units = "in", type = "cairo-png", res = 300)
hc_mtorf_rep
dev.off()
```

```{r}
hc_site <- hc %>% ggtree() %<+% samples + 
  layout_dendrogram() + 
  geom_tippoint(aes(fill = `Region:Site`), shape = 21, color = "black", alpha = 0.8, stroke = 0.5, hjust = 0.3, show.legend = TRUE) +
  # geom_tiplab(size = 2, hjust = 1.05) +
  # scale_fill_manual(values = mtorf_colors, name = "mtORF haplotype") +
  scale_y_continuous(expand = expansion(mult = c(0.02, 0))) +
  # theme_dendrogram(legend.position = c(0.9, 0.82),
  #       legend.margin = margin(0,0,0,0, unit = "mm"),
  #       plot.margin = margin(2,2,2,2, unit = "mm")) +
  xlab("Genetic distance (1-IBS)")

hc_site
```


### Figure 2
```{r}
# pdf("./outputs/figures/Fig2_ibs_dendrogram_species_091423.pdf", width = 6.5, height = 5)
hc_panpopgen %>% ggtree() %<+% samples + 
  layout_dendrogram() +
  # geom_vline(xintercept = -0.15, stroke = 0.5, color = mtorf_colors[1], linetype = 2) +
  # geom_vline(xintercept = -0.05, stroke = 0.5, color = mtorf_colors[2], linetype = 2) +
  # geom_vline(xintercept = -0.105, stroke = 0.5, color = mtorf_colors[3], linetype = 2) +
  # geom_vline(xintercept = -0.103, stroke = 0.5, color = mtorf_colors[4], linetype = 2) +
  # geom_tippoint(aes(fill = `Species`), shape = 21, alpha = 0.8, stroke = 0.5, color = "black", hjust = 0.3, show.legend = TRUE) +
  geom_tippoint(aes(fill = `Species`), shape = 21, alpha = 0.8, stroke = 0.5, color = art.clones, hjust = 0.3) + 
  
  # I WANT TO COLOR THE TIP POINTS BY CLONE GROUPS! THEN ADD NODE POINTS FOR EACH SPP
  
  # geom_tiplab(size = 2, hjust = 1.05) +
  # geom_treescale(width = 0.5)
  theme_dendrogram() +
  scale_fill_manual(values = spp_colors, name = "Species") +
   theme(legend.position = c(0.8, 0.8),
        legend.margin = margin(0,0,0,0, unit = "mm"),
        plot.margin = margin(5,5,5,5, unit = "mm")) +
  ggtitle(expression(paste(italic("Pocillopora"), " identity-by-state clustering dendrogram")), subtitle = "Panama and Clipperton (n= 136)")
# dev.off()
```
```{r}
# color artificial clones red
art.clones <- rep("black", nrow(ma_panpopgen))
art.clones[rownames(ma_panpopgen) %in% replicates]="red"
# inferred clones
inf.clones <- rep("black", nrow(ma_panpopgen))
```

```{r}
pdf("./outputs/figures/galapagos_ibs_clones.pdf", height = 4.5, width = 6)
hc_usfq %>% ggtree() %<+% samples + 
  geom_tippoint(aes(fill = `Species`), shape = 21, alpha = 0.8, stroke = 0.5, color = "black", hjust = 0.3, show.legend = TRUE) +
  geom_tiplab(size = 2, hjust = -0.05) +
  geom_vline(xintercept = -0.15, stroke = 0.5, color = spp_colors[2], linetype = 2) +
  geom_vline(xintercept = -0.11, stroke = 0.5, color = spp_colors[4], linetype = 2) +
  ggplot2::xlim(-0.45, 0.01) +
  geom_treescale(x = -0.45, y = 0.9, width = 0.05) +
  scale_fill_manual(values = spp_colors[c(2,4)], name = "Species") +
   theme(legend.position = c(0.08, 0.8),
        legend.margin = margin(0,0,0,0, unit = "mm"),
        plot.margin = margin(5,5,5,5, unit = "mm")) +
  ggtitle(expression(paste("Galapagos ", italic("Pocillopora"), " (n=", 40, ")", " identity-by-state clustering dendrogram")))
dev.off()
```

```{r}
pdf("./outputs/figures/ibs_dendrogram_coco_mtorf.pdf", width = 6.5, height = 8.5)
tree_title <- expression(paste("Identity-by-state clustering of Isla del Coco ", italic("Pocillopora"), " samples (n=", 26, ")"))
#
# hc$labels <- gsub("_AP3X.*", "", hc$labels)
#
hc %>% ggtree() %<+% samples + 
  # geom_vline(xintercept = -0.15, stroke = 0.5, color = "grey25", linetype = 2) +
  geom_tippoint(aes(fill = `mtorf_type`), shape = 21, alpha = 0.8, stroke = 0.5, color = "black", hjust = 0.3, show.legend = TRUE) +
  geom_tiplab(size = 3, hjust = -0.05) +
  ggplot2::xlim(-0.35, 0.01) +
  geom_treescale(x = -0.35, y = 0.9, width = 0.05) +
  scale_fill_manual(values = mtorf_colors[c(1,3,4)], name = "mtORF Type") +
   theme(legend.position = c(0.08, 0.8),
        legend.margin = margin(0,0,0,0, unit = "mm"),
        plot.margin = margin(5,5,5,5, unit = "mm")) +
  ggtitle(tree_title)
dev.off()
```

## Threshold clonegroup identification
```{r}
# "artificial clones" (technical replicates)
# samples %>% filter(REPLICATE) %>% View()
perlas358 <- c("Perlas_358_AP3XG49", "Perlas_358b_AP3XG49", "Perlas_358c_AP3XG49")
perlas376 <- c("Perlas_376_AP3XG67", "Perlas_376b_AP3XG67", "Perlas_376c_AP3XG67")
coco618 <- c("Coco_618_AP3XF84", "Coco_618b_AP3XF84", "Coco_618c_AP3XF84")
coco653 <- c("Coco_653_AP3XF97", "Coco_653b_AP3XF97", "Coco_653c_AP3XF97")
pan05 <- c("PAN_05",  "PAN_05b", "PAN_05c")
pan10 <- c("PAN_10", "PAN_10b", "PAN_10c")
pan78 <- c("PAN_78", "PAN_78b")
ura51 <- c("URA_51", "URA_51b")
clip19 <- c("Clip_19", "Clip_19b", "Clip_19c")
#
replicates <- c(perlas358, perlas376, coco618, coco653, pan05, pan10, pan78, ura51, clip19)
```
```{r clonegroup_identification}
mafa <- ma
# color artificial clones red
art.clones <- rep("black",nrow(mafa))
#
art.clones[samples$`Sample ID` %in% replicates]="red"
# 
ColorDendrogram(hclust(as.dist(mafa),"ave"), y = art.clones, branchlength=0.05)
```
#### P. grandis clonegroups
```{r}
# subset matrix into species to define clonegroups
ma_pgra <- ma[samples$Species == "P. grandis", samples$Species == "P. grandis"]
# color artificial clones red
art.clones <- rep("black", nrow(ma_pgra))
art.clones[rownames(ma_pgra) %in% replicates]="red"
#
hc_pgra <- hclust(as.dist(ma_pgra),"ave")
#
ColorDendrogram(hc_pgra, y = art.clones, branchlength=0.035)

# sorting samples into clone groups; singletons go into the same "color" group (for plotting) but different "cn" groups (for GLM modeling later)
cc <- cutree(hc_pgra, h = 0.185)
cc1 <- cc

# count number of samples in each clone group
tc <- table(cc)
# identify singletons, set to same color group
singletons <- as.numeric(names(tc)[tc==1])
cc1[cc1 %in% singletons] <- singletons[1]

# assign color sequence to clone groups --> do this after assigning clone groups to all species in summed sequence?
clone_colors <- labels2colors(as.numeric(as.factor(as.numeric(cc1))))
clone_romans <- as.roman(as.numeric(as.factor(as.numeric(cc))))
names(clone_colors) <- names(cc1)
# changing some colors manually for better visibility
clone_colors[clone_colors=="white"]="khaki3"
clone_colors[clone_colors=="black"]="goldenrod"
clone_colors[clone_colors=="turquoise"]="black"
  
# record sample ID and clone group assignments
samples_clonegroups_pgrandis <- tibble(`Sample ID` = names(cc), clone_number = cc, clone_color = clone_colors, clone_code = paste0("Pgra_", clone_color), SINGLETON = ifelse(clone_color == "black", "TRUE", "FALSE"), CLONE = ifelse(clone_color == "black", "FALSE", "TRUE")) %>% arrange(`Sample ID`)

# plot colored dendrogram with inferred clonegroups
# pdf("./outputs/figures/ibs_pgrandis_clonegroups.pdf", width = 6.5, height = 4)
ColorDendrogram(hc_pgra, y = clone_colors, labels = F,branchlength=0.035)
# dev.off()
```
```{r}
# selecting one coral per clone
sel=data.frame()
for (i in unique(samples_clonegroups_pgrandis$clone_number)) {
		sel[i,1:5] = samples_clonegroups_pgrandis[samples_clonegroups_pgrandis$clone_number==i, 1:5] 
}
write(sel$`Sample ID`, "~/Desktop/pgrandis_noclones.txt")
```


#### P. verrucosa 3a clonegroups
```{r}
ma_pver3a <- ma[samples$Species == "P. verrucosa 3a", samples$Species == "P. verrucosa 3a"]
# color artificial clones red
art.clones <- rep("black", nrow(ma_pver3a))
art.clones[rownames(ma_pver3a) %in% replicates]="red"
#
hc_pver3a <- hclust(as.dist(ma_pver3a),"ave")
#
ColorDendrogram(hc_pver3a, y = art.clones, branchlength=0.035)

# sorting samples into clone groups; singletons go into the same "color" group (for plotting) but different "cn" groups (for GLM modeling later)
cc <- cutree(hc_pver3a, h = 0.11)
cc1 <- cc

# count number of samples in each clone group
tc <- table(cc)
# identify singletons, set to same color group
singletons <- as.numeric(names(tc)[tc==1])
cc1[cc1 %in% singletons] <- singletons[1]

# assign color sequence to clone groups
clone_colors <- labels2colors(as.numeric(as.factor(as.numeric(cc1))))
clone_romans <- as.roman(as.numeric(as.factor(as.numeric(cc))))
names(clone_colors) <- names(cc1)
# changing some colors manually for better visibility
clone_colors[clone_colors=="white"]="khaki3"
clone_colors[clone_colors=="black"]="goldenrod"
clone_colors[clone_colors=="turquoise"]="black"
  
# record sample ID and clone group assignments
samples_clonegroups_pverrucosa3a <- tibble(`Sample ID` = names(cc), clone_number = cc, clone_color = clone_colors, clone_code = paste0("Pver3a_", clone_color), SINGLETON = ifelse(clone_color == "black", "TRUE", "FALSE"), CLONE = ifelse(clone_color == "black", "FALSE", "TRUE"))

# plot colored dendrogram with inferred clonegroups
# pdf("./outputs/figures/ibs_pverrucosa3a_clonegroups.pdf", width = 6.5, height = 4)
ColorDendrogram(hc_pver3a, y = clone_colors, labels = F, branchlength=0.035)
# dev.off()
```
```{r}
# selecting one coral per clone
sel=data.frame()
for (i in unique(samples_clonegroups_pverrucosa3a$clone_number)) {
		sel[i,1:5] = samples_clonegroups_pverrucosa3a[samples_clonegroups_pverrucosa3a$clone_number==i, 1:5] 
}
write(sel$`Sample ID`, "./data/pverrucosa3a_noclones_samples.txt")
```
```{r}
# join clonegroup information
samples_all_clones <- bind_rows(samples_clonegroups_pgrandis, samples_clonegroups_pverrucosa3a) %>% 
  select(-SINGLETON, -CLONE) %>% 
  right_join(samples, by = "Sample ID") %>% relocate(`clone_number`:`clone_code`, .after = `Species`)

write_excel_csv("~/Desktop/samples_all_clonegroups.csv")
```
```{r}
# find top samples within clonegroups and species for phylogenetic reconstructions

samples_clones_subspp_qcdata <- samples_all_clones %>% left_join(., pgrandis_subspp, by = "Sample ID") %>% left_join(., multiqc_data, , by = "Sample ID")
```


### Panama manuscript samples
```{r}
pdf("./outputs/figures/FigS2_ibs_clones.pdf", width = 6.5, height = 6)
# subset matrix into species to define clonegroups
ma_panpopgen <- ma[samples$PANAMA_MANUSCRIPT, samples$PANAMA_MANUSCRIPT]
# re-clustering
hc_panpopgen <- hclust(as.dist(ma_panpopgen),"ave")
plot(hc_panpopgen, cex=0.4)
abline(h=0.145,col="darkorange",lty=3)
abline(h=0.105,col="turquoise2",lty=3)

# color artificial clones red
art.clones <- rep("black", nrow(ma_panpopgen))
art.clones[rownames(ma_panpopgen) %in% replicates]="red"
#
hc_panpopgen <- hclust(as.dist(ma_panpopgen),"ave")
#
ColorDendrogram(hc_panpopgen, y = art.clones, branchlength=0.035)

# sorting samples into clone groups; singletons go into the same "color" group (for plotting) but different "cn" groups (for GLM modeling later)
cc <- cutree(hc_panpopgen, h = 0.145)
cc1 <- cc

# count number of samples in each clone group
tc <- table(cc)
# identify singletons, set to same color group
singletons <- as.numeric(names(tc)[tc==1])
cc1[cc1 %in% singletons] <- singletons[1]

# assign color sequence to clone groups
clone_colors <- labels2colors(as.numeric(as.factor(as.numeric(cc1))))
names(clone_colors) <- names(cc1)
# changing some colors manually for better visibility
clone_colors[clone_colors=="white"]="khaki3"
clone_colors[clone_colors=="black"]="goldenrod"
clone_colors[clone_colors=="turquoise"]="black"
  
# record sample ID and clone group assignments
samples_clonegroups_panama_manuscript <- tibble(`Sample ID` = names(cc), clone_number = cc, clone_color = clone_colors, SINGLETON = ifelse(clone_color == "black", "TRUE", "FALSE"))
  
ColorDendrogram(hc_panpopgen, y = clone_colors, labels = F, branchlength=0.035)
dev.off()
```

## Plot IBS clustering with clones removed
```{r}
ma <- as.matrix(read.table("./outputs/angsd/full_noclones_ibs05.ibsMat"))
```
```{r bamfile_sample_order_check}
# reading list of bam files = order of samples in IBS matrix
bams <- read.table("/Users/mikeconnelly/computing/projects/etp_pocillopora_gskim/outputs/angsd/full_noclones_bamfile.txt",header=F)[,1]
#
bam_ids <- sub("/scratch/nmnh_corals/connellym/projects/etp_pocillopora_gskim/outputs/alignments/", "", bams, perl=T)
bam_ids <- sub(".sorted.md.rg.bam", "", bam_ids, perl=T)
bam_ids <- sub("_S[[:digit:]]+$", "", bam_ids)
#
bam_ids
```
```{r}
# aligning sample metadata with bams list, then check again
samples_noclones <- samples[match(bam_ids, samples$`Sample ID`),] 
# do all the sample ids match the bam file ids in the correct order?
all(samples_noclones$`Sample ID` == bam_ids) # TRUE
```
```{r}
# IMPORTANT: verify that the order of ids in sample metadata and in the bam file list used for IBS in ANGSD match
dimnames(ma) <- list(samples_noclones$`Sample ID`, samples_noclones$`Sample ID`)
```
```{r}
# plotting hierarchical clustering dendrogram
hc <- hclust(as.dist(ma), "ave")
plot(hc, cex = 0.4)
```
```{r}
pdf("./outputs/figures/sicb/ibs_dendrogram_all_noclones_mtorf.pdf", width = 8.5, height = 4.85)
hc_noclones <- hc %>% ggtree() %<+% samples + 
  layout_dendrogram() + 
  geom_tippoint(aes(fill = `mtorf_type`), shape = 21, color = "black", alpha = 0.8, stroke = 0.5, hjust = 0.3, show.legend = T) +
  # geom_tiplab(size = 2, hjust = 1.05) +
  scale_fill_manual(values = c(mtorf_colors), name = "mtORF haplotype") +
  scale_y_continuous(expand = expansion(mult = c(0.02, 0))) +
  theme_dendrogram(legend.position = c(0.9, 0.82),
        legend.margin = margin(0,0,0,0, unit = "mm"),
        plot.margin = margin(2,2,2,2, unit = "mm")) +
  xlab("Genetic distance (1-IBS)")
dev.off()
```

```{r}
png("./outputs/figures/sicb/ibs_dendrogram_all_noclones_mtorf.png", width = 8.5, height = 4.85, units = "in", type = "cairo-png", res = 300)
hc_noclones
dev.off()
```

```{r}
# histogram of pairwise distances 

# need to select only columns of samples 
ma_type1a <- madf %>% filter(mtorf_type == "Type 1a") %>% select(.$`Sample ID`) %>% as.matrix() %>% hist()

hist(madf[madf$mtorf_type == "Type 1a", madf$mtorf_type == "Type 1a"])
```

