---
title: "Pocillopora mtORF analysis"
author: "Mike Connelly"
date: "6/7/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```
```{r packages, message=FALSE}
library("tidyverse")
library("ape")
library("pegas")
#
source("./R/pocillopora_etp_gksim_functions.R")
```

## Import sample metadata
```{r import_sample_metadata}
source("./R/sample_import.R")
samples <- all_samples
# NOTE: Perform mtORF barcode extraction and alignment for all samples before clone correction
```

## Import mtORF sequence alignment 
```{r import_alignment}
data <- read.dna("./outputs/mitofinder/mtorf_hits_aligned.fasta", format = "fasta")
checkAlignment(data, what = 1:4)
# 325 sequences, 829 sites, no gaps
# attributes(data)
# dimnames(data)[[1]]
```
```{r haplotypes}
dataHaplo <- haplotype(data)
dataHaplo
```
```{r extract haplotypes and build network}
dataHaplo <- sort(dataHaplo, what = "labels")
# attributes(dataHaplo)
dataNet <- haploNet(dataHaplo)
# dataNet <- mst(dataHaplo)
# dataNet <- mjn(dataHaplo)
```
```{r extract sequence labels}
seq_labels <- dimnames(data)[[1]]
# seq_labels
seq_ids <- gsub("_S[[:digit:]]+.[[:digit:]]$", "", seq_labels)
# seq_ids
```
```{r countHap}
countHap <- function(hap = h, dna = x){
  with(
    stack(setNames(attr(hap, "index"), rownames(hap))),
    table(haplotype = ind, seq = attr(dna, "dimnames")[[1]][values])
  )
}
```
```{r genotype haplotype matches}
df <- as.data.frame(countHap(dataHaplo, data))
good <- df %>% filter(Freq == 1)
```
```{r select_good_haplotypes}
good$`Sample ID` <- gsub("_S.*", "", good$seq)
#
sample_haplotypes <- good %>% dplyr::select(`Sample ID`, `haplotype`)
```
```{r mtORF names}
# correspondence table between identified haplotypes and mtORF type/ORF name, etc.
mtorf_types <- c(I = "Type 3a", II = "Type 3b", III = "Type3d", IV = "Type 1a", V = "Type 2")
ORF_types <- c(I = "ORF46", II = "ORF47", III = "ORF54", IV = "ORF27", V = "ORF01")
#
sample_haplotypes$mtorf_type <-  mtorf_types[sample_haplotypes$haplotype]
```
```{r match samples}
## need sample order to match to order of aligned sequences in FASTA file
samples_fasta <- samples[match(seq_ids, samples$`Sample ID`),]
```
```{r missing mtorf sample correction}
# identify samples without recovered mtORF sequences 
samples_nomtorf <- samples %>% filter(`Sample ID` %notin% good$`Sample ID`)
  
# filter out samples without recovered mtORF sequences
samples_mtorf <- samples_fasta %>% filter(samples_fasta$`Sample ID` %in% good$`Sample ID`)

# filter out clonal replicates identified by IBS
# samples_noclones
```

### Plot haplotype networks
```{r basic haplotype net}
plot(dataNet)
```
```{r blank haplotype net}
# note: if scaling size by number of sequences, correct for clones first, as mtORF type 1 is clearly dominant.
# pdf(file="~/Desktop/mtORF_haplotypes_sites.pdf", width = 8, height = 10, pointsize = 10)
plot(dataNet,
     size = attr(dataNet, "freq"),
     show.mutation = 3,
     labels = F,
     legend = c(-100,100))
# dev.off()
```
```{r metadata haplotype frequencies}
Region <- haploFreq(data, samples_mtorf$Region, haplo = dataHaplo)
Site <- haploFreq(data, samples_mtorf$Site, haplo = dataHaplo)
Type <- haploFreq(data, samples_mtorf$mtorf_type, haplo = dataHaplo)
Species <- haploFreq(data, samples_mtorf$Species, haplo = dataHaplo)
NGSAdmix_Species <- haploFreq(data, samples_mtorf$Species, haplo = dataHaplo)
HaploType <- haploFreq(data, samples_mtorf$haplotype, haplo = dataHaplo)
```
```{r colored haplonet}
pdf(file="./outputs/figures/mtORF_haplotypes_region.pdf", width = 4, height = 5, pointsize = 6)
plot(dataNet,
     size = attr(dataNet, "freq"),
     pie = Region,
     bg = c("blue", "green", "yellow", "red"),
     show.mutation = 1,
     labels = T,
     legend = c(-3,18))
dev.off()
```

### Figure S4
```{r}
setHaploNetOptions(mutations.font = 20)
```
```{r figS4A_mtorf_haplonet}
pdf(file="./outputs/figures/FigS4A_mtORF_haplotypes_type.pdf", width = 6.5, height = 3.5, pointsize = 6)
blank_col = rep("white", 5)
plot(dataNet,
     size = sqrt(attr(dataNet, "freq")),
     pie = Type,
     # bg = blank_col, 
     bg = c("darkorange1", "purple", "turquoise2", "turquoise4", "darkslategrey"),
     show.mutation = 1, 
     labels = FALSE,
     legend = FALSE)
dev.off()
```
```{r}
pdf(file="./outputs/figures/mtORF_haplotypes_species.pdf", width = 4, height = 2, pointsize = 6)
plot(dataNet,
     size = sqrt((attr(dataNet, "freq"))),
     pie = Species,
     bg = spp_colors,
     show.mutation = 1,
     labels = FALSE,
     legend = FALSE)
dev.off()
```
```{r figS4B}
pdf("./outputs/figures/FigS4B_mtORF_alignment.pdf", width = 6.5, height = 3.5)
checkAlignment(data, what = 4)
dev.off()
```

## Subset to no-clones samples
```{r}
# data
# labels(data)
```
```{r}
table(str_replace_all(dimnames(data)[[1]], "_S.*$", "") %in% samples_noclones_ngsadmix$`Sample ID`)

data_noclones <- data[str_replace_all(dimnames(data)[[1]], "_S.*$", "") %in% samples_noclones_ngsadmix$`Sample ID`, ]
```
```{r}
checkAlignment(data_noclones)
```
```{r haplotypes}
dataHaplo <- haplotype(data_noclones)
dataHaplo
```
```{r extract haplotypes and build network}
dataHaplo <- sort(dataHaplo, what = "labels")
# attributes(dataHaplo)
dataNet <- haploNet(dataHaplo)
# dataNet <- mst(dataHaplo)
# dataNet <- mjn(dataHaplo)
```
```{r extract sequence labels}
seq_labels <- dimnames(data_noclones)[[1]]
# seq_labels
seq_ids <- gsub("_S[[:digit:]]+.[[:digit:]]$", "", seq_labels)
# seq_ids
```
```{r countHap}
countHap <- function(hap = h, dna = x){
  with(
    stack(setNames(attr(hap, "index"), rownames(hap))),
    table(haplotype = ind, seq = attr(dna, "dimnames")[[1]][values])
  )
}
```
```{r genotype haplotype matches}
df <- as.data.frame(countHap(dataHaplo, data_noclones))
good <- df %>% filter(Freq == 1)
```
```{r select_good_haplotypes}
good$`Sample ID` <- gsub("_S.*", "", good$seq)
#
sample_haplotypes <- good %>% dplyr::select(`Sample ID`, `haplotype`)
```
```{r mtORF names}
# correspondence table between identified haplotypes and mtORF type/ORF name, etc.
mtorf_types <- c(I = "Type 3a", II = "Type 3b", III = "Type3d", IV = "Type 1a", V = "Type 2")
ORF_types <- c(I = "ORF46", II = "ORF47", III = "ORF54", IV = "ORF27", V = "ORF01")
#
sample_haplotypes$mtorf_type <-  mtorf_types[sample_haplotypes$haplotype]
```
```{r match samples}
## need sample order to match to order of aligned sequences in FASTA file
samples_fasta <- samples_noclones_ngsadmix[match(seq_ids, samples_noclones_ngsadmix$`Sample ID`),]
```
## Plot haplotype networks
```{r basic haplotype net}
plot(dataNet)
```
```{r blank haplotype net}
# note: if scaling size by number of sequences, correct for clones first, as mtORF type 1 is clearly dominant.
# pdf(file="~/Desktop/mtORF_haplotypes_sites.pdf", width = 8, height = 10, pointsize = 10)
plot(dataNet,
     size = attr(dataNet, "freq"),
     show.mutation = 3,
     labels = F,
     legend = c(-100,100))
# dev.off()
```
```{r metadata haplotype frequencies}
Region <- haploFreq(data_noclones, samples_fasta$Region, haplo = dataHaplo)
Site <- haploFreq(data_noclones, samples_fasta$Site, haplo = dataHaplo)
Type <- haploFreq(data_noclones, samples_fasta$mtorf_type, haplo = dataHaplo)
Species <- haploFreq(data_noclones, samples_fasta$Species, haplo = dataHaplo)
NGSAdmix_Species <- haploFreq(data_noclones, samples_fasta$`NGSAdmix K6 Species`, haplo = dataHaplo)
HaploType <- haploFreq(data_noclones, samples_fasta$haplotype, haplo = dataHaplo)
```
```{r colored haplonet}
pdf(file="./outputs/figures/mtORF_haplotypes_region.pdf", width = 4, height = 5, pointsize = 6)
plot(dataNet,
     size = attr(dataNet, "freq"),
     pie = Region,
     bg = c("blue", "green", "yellow", "red"),
     show.mutation = 1,
     labels = T,
     legend = c(-3,18))
dev.off()
```
```{r mtorf_haplonet_type_noclones}
plot(dataNet,
     size = sqrt(attr(dataNet, "freq")),
     pie = Type,
     bg = c("darkorange1", "purple", "turquoise2", "turquoise4", "darkslategrey"),
     show.mutation = 1, 
     labels = FALSE,
     legend = FALSE)
```
```{r figS4C_mtorf_haplonet_ngsadmixK6}
pdf(file="./outputs/figures/FigS4C_mtORF_haplotypes_spp.pdf", width = 6.5, height = 3.5, pointsize = 6)
plot(dataNet,
     size = sqrt(attr(dataNet, "freq")),
     pie = NGSAdmix_Species,
     bg = ngsadmix_pop_colors_k6[c(1,4,3,2,5,6,7)],
     show.mutation = 1, 
     labels = FALSE,
     legend = FALSE)
dev.off()
```
```{r figS4B}
pdf("./outputs/figures/FigS4D_mtORF_alignment.pdf", width = 6.5, height = 3.5)
checkAlignment(data_noclones, what = 4)
dev.off()
```
