---
title: "PCAngsd"
author: "Mike Connelly"
date: "2023-05-10"
output: html_document
---
## Setup and packages
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, cache = TRUE)
options(stringsAsFactors = FALSE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```
```{r packages}
library("tidyverse")
library("ggrepel")
#
source("./R/pocillopora_etp_gksim_functions.R")
```

## Import sample metadata
```{r import_sample_metadata}
source("./R/sample_import.R")
samples <- all_samples
```

NOTE: Remove clones from the dataset using IBS analysis before running PCAngsd. Code for selecting no-clones samples based on highest proportion of sites w/ 5x coverage is in 02_pipeline_metrics.Rmd

## All samples (clones included) PCA
```{r pca}
# *P. grandis* aligments of full analysis samples
C <- as.matrix(read.table("./outputs/angsd/full_analysis_pgra_ibs05.cov"))
e <- eigen(C)
```
```{r pca_test_plot}
plot(e$vectors[,1:2],xlab="PC1",ylab="PC2", main="individual allele frequency")
#
plot(e$vectors[,2:3],xlab="PC2",ylab="PC3", main="individual allele frequency")
# 
```
```{r combine_pca_metadata}
# combine pc coordinates with sample metadata
pcs <- as.data.frame(e$vectors) %>% cbind(samples)
# store variable with % variance explained by each PC
pcvar <- round(e$values, 2)
#
```
### Plot PCA of all-samples
```{r pca_test_ggplot}
title <- expression(paste("PCA of East Pacific ", italic("Pocillopora"), " samples"))
# test ggplot 
pcs %>% ggplot(aes(x = V1, y = V2)) +
  geom_point(aes(fill = `Region`), shape = 21, size = 2, alpha = 0.75) +
  scale_fill_manual(values = region_colors) +
  # facet_grid(Region ~ .) +
  coord_fixed(pcvar[2]/pcvar[1]) + 
  xlab(paste0( "PC1 (", pcvar[1], "%)")) + 
  ylab(paste0( "PC2 (", pcvar[2], "%)")) +
  ggtitle(title)
# 
pcs %>% ggplot(aes(x = V1, y = V2)) +
  geom_point(aes(fill = `mtorf_type`), shape = 21, size = 2, alpha = 0.75) +
  # scale_fill_manual(values = region_colors) +
  # facet_grid(Region ~ .) +
  coord_fixed(pcvar[2]/pcvar[1]) +
  xlab(paste0( "PC1 (", pcvar[1], "%)")) +
  ylab(paste0( "PC2 (", pcvar[2], "%)")) +
  ggtitle(title)
# 
pcs %>% ggplot(aes(x = V1, y = V2)) +
  geom_point(aes(fill = `Species`), shape = 21, size = 2, alpha = 0.75) +
  scale_fill_manual(values = spp_colors) +
  # scale_shape_manual(values = c(21:25)) +
  # facet_grid(Region ~ .) +
  coord_fixed(pcvar[2]/pcvar[1]) + 
  xlab(paste0( "PC1 (", pcvar[1], "%)")) + 
  ylab(paste0( "PC2 (", pcvar[2], "%)")) +
  ggtitle(title)
# 
```
```{r figure 1: pca_1-2}
# use custom function to plot PCA colored by Region:Site
pdf("./outputs/figures/pca12.pdf", width = 8, height = 6)
gg_pcangsd(e, samples, pc=c(1,2)) +
  ggtitle(title)
  #
gg_pcangsd(e, samples, pc=c(1,2)) +
  facet_grid(Region ~ .) +
  ggtitle(title)
dev.off()
```
```{r pca_2-3}
pdf("./outputs/figures/pca23.pdf", width = 8, height = 6)
gg_pcangsd(e, samples, pc=c(2,3)) +
  ggtitle(title)
  #
dev.off()
```

## No-clones PCA (w/ linked SNPs)
```{r pca}
# *P. grandis* aligments of no-clones samples
C <- as.matrix(read.table("./outputs/angsd/final_noclones_pgra_ibs05.cov"))
e <- eigen(C)
```
```{r pca_test_plot}
plot(e$vectors[,c(1,2)],xlab="PC1",ylab="PC2", main="individual allele frequency")
#
plot(e$vectors[,c(2,3)],xlab="PC2",ylab="PC3", main="individual allele frequency")
#
plot(e$vectors[,c(3,4)],xlab="PC3",ylab="PC4", main="individual allele frequency")
```

```{r bamfile_sample_order_check}
# reading list of bam files = order of samples in IBS matrix
bams <- read.table("/Users/mikeconnelly/computing/projects/etp_pocillopora_gskim/outputs/angsd/final_noclones_bamfile.txt",header=F)[,1]
#
bam_ids <- sub("/scratch/nmnh_corals/connellym/projects/etp_pocillopora_gskim/outputs/alignments/", "", bams, perl=T)
bam_ids <- sub(".sorted.md.rg.bam", "", bam_ids, perl=T)
bam_ids <- sub("_S[[:digit:]]+$", "", bam_ids)
#
bam_ids
```
```{r select_samples}
# aligning sample metadata with bams list, then check again
samples_noclones <- samples[match(bam_ids, samples$`Sample ID`),] 
# do all the sample ids match the bam file ids in the correct order?
all(samples_noclones$`Sample ID` == bam_ids) # TRUE
```
```{r combine_pca_metadata}
# combine pc coordinates with sample metadata
pcs <- as.data.frame(e$vectors) %>% cbind(samples_noclones_ngsadmix) %>% left_join(samples)
# store variable with % variance explained by each PC
pcvar <- round(e$values, 2)
#
```
```{r screeplot}
pcvardf <- data.frame("PC"=seq(1:length(pcvar)), pcvar)
pcvardf %>% ggplot(aes(PC, pcvar)) + geom_col() + xlim(c(0,10))
# only first 4 PCs have >5% variance explained
```
```{r}
# pdf(file="./outputs/figures/sicb/pcangsd_species.pdf", width = 5.5, height = 3.75)

pca <- pcs %>% ggplot(aes(x = V1, y = V2)) +
  geom_point(aes(fill = `Species`), shape = 21, size = 3, alpha = 0.75, show.legend = F) +
  scale_fill_manual(values = spp_colors) +
  # facet_grid(Region ~ .) +
  # facet_wrap(mtorf_type ~ .) +
  coord_fixed(pcvar[2]/pcvar[1]) +
  xlab(paste0( "PC1 (", pcvar[1], "%)")) + 
  ylab(paste0( "PC2 (", pcvar[2], "%)")) +
  theme(plot.title = element_text(size = 16, hjust = -0.2),
        axis.title = element_text(size = 14),
        axis.text = element_text(size = 12),
        legend.title = element_text(size = 14),
        legend.text = element_text(size = 12)) 

print(pca)
# dev.off()
```

### PCA of all samples by species
```{r}
# pdf(file="./outputs/figures/etp_spp_pca_color.pdf", width = 4, height = 2.5)
pcs %>% ggplot(aes(x = V1, y = V2)) +
  geom_jitter(aes(fill = `Species`), shape = 21, size = 4, alpha = 0.6, show.legend = T, width = 0.005, height = 0.005) +
  scale_fill_manual(values = spp_colors) +
  # facet_wrap(Region ~ .) +
  coord_fixed(pcvar[2]/pcvar[1]) +
  xlab(paste0( "PC1 (", pcvar[1], "%)")) + 
  ylab(paste0( "PC2 (", pcvar[2], "%)")) +
  theme(plot.title = element_text(size = 16, hjust = 1.4),
        axis.title = element_text(size = 12),
        axis.text = element_text(size = 11),
        legend.title = element_text(size = 14),
        legend.text = element_text(size = 12, face = "italic")) +
  ggtitle(title)
# dev.off()
```

### Figure S6
```{r figS6_pca_linked}
pdf(file="./outputs/figures/FigS6_pcangsd_linked_ngsadmix_species.pdf", width = 6.65, height = 2.25)
#
title <- expression(paste("East Pacific ", italic("Pocillopora"), " PCA (5,587,480 SNPs)"))
#
pca <- pcs %>%
  mutate("Region" = factor(`Region`, levels = region_order_longitude, ordered = T),
         "NGSAdmix K6 Species" = factor(`NGSAdmix K5 Species`, levels = ngsadmix_spp, ordered = T)) %>% 
  ggplot(aes(x = V1, y = V2)) +
  geom_point(aes(fill = `NGSAdmix K5 Species`), shape = 21, size = 4, alpha = 0.65, show.legend = T) +
  scale_fill_manual(values = ngsadmix_pop_colors, name = "NGSAdmix Species (K=5)") + 
  guides(fill = guide_legend(override.aes = list(size = 3, alpha = 1))) +
  coord_fixed(pcvar[2]/pcvar[1]) +
  xlab(paste0( "PC1 (", pcvar[1], "%)")) + 
  ylab(paste0( "PC2 (", pcvar[2], "%)")) +
  guides(fill=guide_legend(nrow=6,byrow=TRUE)) +
  theme(plot.title = element_text(size = 16, hjust = -0.2),
        axis.title = element_text(size = 12),
        axis.text = element_text(size = 10),
        legend.title = element_text(size = 10),
        legend.text = element_text(size = 8),
        legend.position = "none",
        legend.key.spacing.y = unit(-1, "mm"))
         
print(pca)
dev.off()
```

## No-clones PCA (unlinked SNPs after LD pruning)
```{r pca}
# *P. grandis* aligments
C <- as.matrix(read.table("./outputs/angsd/final_noclones/final_noclones_noLD_pgra_himb_filtered.cov"))
e <- eigen(C)
```
```{r pca_test_plot}
plot(e$vectors[,c(1,2)],xlab="PC1",ylab="PC2", main="individual allele frequency")
#
plot(e$vectors[,c(2,3)],xlab="PC2",ylab="PC3", main="individual allele frequency")
#
plot(e$vectors[,c(3,4)],xlab="PC3",ylab="PC4", main="individual allele frequency")
```
```{r bamfile_sample_order_check}
# reading list of bam files = order of samples in IBS matrix
bams <- read.table("./outputs/angsd/final_noclones/final_noclones_pgra_bamfile.txt",header=F)[,1]
#
bam_ids <- sub("/scratch/nmnh_corals/connellym/projects/etp_pocillopora_gskim/outputs/alignments/pgra_himb/", "", bams, perl=T)
bam_ids <- sub(".sorted.md.rg.bam", "", bam_ids, perl=T)
bam_ids <- sub("_S[[:digit:]]+$", "", bam_ids)
#
bam_ids
```
```{r select_samples}
# aligning sample metadata with bams list, then check again
samples_noclones_ngsadmix <- samples_noclones_ngsadmix[match(bam_ids, samples_noclones_ngsadmix$`Sample ID`),] 
# do all the sample ids match the bam file ids in the correct order?
all(samples_noclones_ngsadmix$`Sample ID` == bam_ids) # TRUE
```
```{r combine_pca_metadata}
# combine pc coordinates with sample metadata
pcs <- as.data.frame(e$vectors) %>% cbind(samples_noclones_ngsadmix) %>% left_join(samples)
# store variable with % variance explained by each PC
pcvar <- round(e$values, 2)
#
```
```{r screeplot}
pcvardf <- data.frame("PC"=seq(1:length(pcvar)), pcvar)
pcvardf %>% ggplot(aes(PC, pcvar)) + geom_col() + xlim(c(0,10))
# only first 4 PCs have >5% variance explained
```
```{r pca_visualization}
# pdf(file="./outputs/figures/sicb/pcangsd_species.pdf", width = 5.5, height = 3.75)
pca <- pcs %>% ggplot(aes(x = V1, y = V2)) +
  geom_point(aes(fill = `NGSAdmix K5 Species`), shape = 21, size = 3, alpha = 0.75, show.legend = T) +
  scale_fill_manual(values = ngsadmix_pop_colors_k6) +
  # facet_grid(Region ~ .) +
  # facet_wrap(mtorf_type ~ .) +
  coord_fixed(pcvar[2]/pcvar[1]) +
  xlab(paste0( "PC1 (", pcvar[1], "%)")) + 
  ylab(paste0( "PC2 (", pcvar[2], "%)")) +
  theme(plot.title = element_text(size = 16, hjust = -0.2),
        axis.title = element_text(size = 14),
        axis.text = element_text(size = 12),
        legend.title = element_text(size = 14),
        legend.text = element_text(size = 12)) 
print(pca)
# dev.off()
```
```{r pca_2_3}
pca <- pcs %>% ggplot(aes(x = V2, y = V3)) +
  geom_point(aes(fill = `NGSAdmix K5 Species`), shape = 21, size = 3, alpha = 0.75, show.legend = F) +
  scale_fill_manual(values = ngsadmix_pop_colors_k6) +
  # facet_grid(Region ~ .) +
  # facet_wrap(mtorf_type ~ .) +
  coord_fixed(pcvar[3]/pcvar[2]) +
  xlab(paste0( "PC2 (", pcvar[2], "%)")) + 
  ylab(paste0( "PC3 (", pcvar[3], "%)")) +
  theme(plot.title = element_text(size = 16, hjust = -0.2),
        axis.title = element_text(size = 14),
        axis.text = element_text(size = 12),
        legend.title = element_text(size = 14),
        legend.text = element_text(size = 12))
print(pca)
```
```{r pca_2_4}
pca <- pcs %>% ggplot(aes(x = V2, y = V4)) +
  geom_point(aes(fill = `NGSAdmix K5 Species`), shape = 21, size = 3, alpha = 0.75, show.legend = F) +
  scale_fill_manual(values = ngsadmix_pop_colors_k6) +
  # facet_grid(Region ~ .) +
  # facet_wrap(mtorf_type ~ .) +
  coord_fixed(pcvar[4]/pcvar[2]) +
  xlab(paste0( "PC3 (", pcvar[2], "%)")) + 
  ylab(paste0( "PC4 (", pcvar[4], "%)")) +
  theme(plot.title = element_text(size = 16, hjust = -0.2),
        axis.title = element_text(size = 14),
        axis.text = element_text(size = 12),
        legend.title = element_text(size = 14),
        legend.text = element_text(size = 12))
print(pca)
```

### Figure 1A
```{r fig1A_pca}
pdf(file="./outputs/figures/Fig1A_pcangsd_ngsadmix_species.pdf", width = 6.65, height = 3.25)
#
title <- expression(paste("East Pacific ", italic("Pocillopora"), " PCA (1,310,435 unlinked SNPs)"))
#
pca <- pcs %>%
  mutate("Region" = factor(`Region`, levels = region_order_longitude, ordered = T),
         "NGSAdmix K6 Species" = factor(`NGSAdmix K5 Species`, levels = ngsadmix_spp, ordered = T)) %>% 
  ggplot(aes(x = V1, y = V2)) +
  geom_point(aes(fill = `NGSAdmix K5 Species`), shape = 21, size = 4, alpha = 0.65, show.legend = T) +
  scale_fill_manual(values = ngsadmix_pop_colors, name = "NGSAdmix Species (K=5)") + 
  guides(fill = guide_legend(override.aes = list(size = 3, alpha = 1))) +
  coord_fixed(pcvar[2]/pcvar[1]) +
  xlab(paste0( "PC1 (", pcvar[1], "%)")) + 
  ylab(paste0( "PC2 (", pcvar[2], "%)")) +
  theme(plot.title = element_text(size = 16, hjust = -0.2),
        legend.background = element_rect(color = "lightgrey"),
        # panel.grid = element_blank(),
        axis.title = element_text(size = 12),
        axis.text = element_text(size = 10),
        legend.title = element_text(size = 10, margin = margin(b = 0.5, unit = "mm")),
        legend.text = element_text(size = 8),
        legend.position = "none",
        # legend.position = c(0.151, 0.695),
        legend.key.spacing.y = unit(-1.6, "mm"))
         
print(pca)
dev.off()
```

## Perform separate PCAs for different species
### Plot PCAs for *P. grandis* - no clones, offshore vs. continental selection
```{r pca_c1}
# C1 <- as.matrix(read.table("./outputs/angsd/pgrandis_noclones.cov"))
C1 <- as.matrix(read.table("./outputs/angsd/pgrandiscontinent_pgrandisoffshore_ibs05.cov"))
C1 <- as.matrix(read.table("./outputs/angsd/pgrandiscontinent_pgrandisoffshore_noLD.cov"))
e1 <- eigen(C1)
```
```{r pca_test_plot}
plot(e1$vectors[,1:2],xlab="PC1",ylab="PC2", main="individual allele frequency")
#
plot(e1$vectors[,2:3],xlab="PC2",ylab="PC3", main="individual allele frequency")
# 
```
```{r bamfile_sample_order_check}
# reading list of bam files = order of samples in IBS matrix
bams <- read.table("/Users/mikeconnelly/computing/projects/etp_pocillopora_gskim/outputs/angsd/pgrandiscontinent_pgrandisoffshore_pgra_bamfile.txt",header=F)[,1]
#
bam_ids <- sub("/scratch/nmnh_corals/connellym/projects/etp_pocillopora_gskim/outputs/alignments/", "", bams, perl=T)
bam_ids <- sub("pgra_himb/", "", bam_ids, perl=T)
bam_ids <- sub(".sorted.md.rg.bam", "", bam_ids, perl=T)
bam_ids <- sub("_S[[:digit:]]+$", "", bam_ids)
#
# bam_ids
```
```{r select_samples}
# aligning sample metadata with bams list, then check again
samples_pgrandis <- samples_noclones_ngsadmix[match(bam_ids, samples_noclones_ngsadmix$`Sample ID`),] 
# do all the sample ids match the bam file ids in the correct order?
all(samples_pgrandis$`Sample ID` == bam_ids) # TRUE
```
```{r combine_pca_metadata}
# combine pc coordinates with sample metadata
pcs_c1 <- as.data.frame(e1$vectors) %>% cbind(samples_pgrandis)
# store variable with % variance explained by each PC
pcvar_c1 <- round(e1$values, 2)
#
```
```{r pca_metadata_factors}
pcs_c1$Region <- factor(pcs_c1$Region, levels = c("Panama", "Chiriqui", "Golfo Dulce", "Bahia Culebra", "Gorgona", "Coco", "Galapagos Central", "Galapagos Northern", "Clipperton"), ordered = T)
```
```{r pca_ggplot}
# test ggplot 
pcs_c1 %>% ggplot(aes(x = V1, y = V2)) +
  geom_point(aes(fill = `Region`), shape = 21, size = 2, alpha = 0.75) +
  scale_fill_manual(values = region_colors) +
  facet_wrap(~ `Region:Site`) +
  coord_fixed(pcvar_c1[2]/pcvar_c1[1]) + 
  theme(legend.position = "right") +
  xlab(paste0( "PC1 (", pcvar_c1[1], "%)")) + 
  ylab(paste0( "PC2 (", pcvar_c1[2], "%)"))
# 
# test ggplot 
pcs_c1 %>% ggplot(aes(x = V1, y = V2)) +
  geom_point(aes(fill = `Region:Site`), shape = 21, size = 2, alpha = 0.75) +
  # scale_fill_manual(values = site_colors) +
  # facet_grid(Region ~ .) +
  coord_fixed(pcvar_c1[2]/pcvar_c1[1]) + 
  theme(legend.position = "right") +
  xlab(paste0( "PC1 (", pcvar_c1[1], "%)")) + 
  ylab(paste0( "PC2 (", pcvar_c1[2], "%)"))
# 
```

### Figure ###
```{r pgrandis_region_pca}
png("./outputs/figures/Evolution2024_pgrandis_region_pca.png", height = 3, width = 4.5, type = "cairo-png", unit = "in", res = 300)
title <- expression(paste("PCA of ETP ", italic("P. grandis"), " samples (n = 163)"))
#
pcs_c1 %>%
  filter(`NGSAdmix K5 Species` %notin% c("Unassigned", NA)) %>% 
  ggplot(aes(x = V1, y = V2)) +
  geom_point(aes(fill = `NGSAdmix K5 Species`), shape = 21, size = 2, alpha = 0.75) +
  scale_fill_manual(values = c("darkorange3", "darkorange", "grey90")) +
  # facet_grid(Region ~ .) +
  coord_fixed(pcvar_c1[2]/pcvar_c1[1]) + 
  theme(legend.position = "none") +
  xlab(paste0( "PC1 (", pcvar_c1[1], "%)")) + 
  ylab(paste0( "PC2 (", pcvar_c1[2], "%)"))
# 
dev.off()
```

### Plot PCAs for P. grandis - Panama
```{r pca_c1}
pgrandis <- read_csv("./data/pgrandis_panama_samples.txt", col_names = "Sample ID") %>% arrange(`Sample ID`)
pgrandis_samples <- pgrandis %>% left_join(samples)
#
C1 <- as.matrix(read.table("./outputs/angsd/pgrandis_panama.cov"))
e1 <- eigen(C1)
```
```{r pca_test_plot}
plot(e1$vectors[,1:2],xlab="PC1",ylab="PC2", main="individual allele frequency")
#
plot(e1$vectors[,2:3],xlab="PC2",ylab="PC3", main="individual allele frequency")
# 
```
```{r combine_pca_metadata}
# combine pc coordinates with sample metadata
pcs_c1 <- as.data.frame(e1$vectors) %>% cbind(pgrandis_samples)
# store variable with % variance explained by each PC
pcvar_c1 <- round(e1$values, 2)
#
```
```{r pca_ggplot}
# test ggplot 
pdf("~/Desktop/pgrandis_subspecies.pdf", width = 8.5)
pcs_c1 %>% ggplot(aes(x = V1, y = V2)) +
  geom_point(aes(fill = `Country`), shape = 21, size = 2, alpha = 0.75) +
  geom_text_repel(aes(label = `Sample ID`), size = 1, max.overlaps = 50) +
  # scale_fill_manual(values = region_colors) +
  # facet_grid(Region ~ .) +
  coord_fixed(pcvar_c1[2]/pcvar_c1[1]) + 
  theme(legend.position = "right") +
  xlab(paste0( "PC1 (", pcvar_c1[1], "%)")) + 
  ylab(paste0( "PC2 (", pcvar_c1[2], "%)"))
dev.off()
# 
# test ggplot 
pcs_c1 %>% ggplot(aes(x = V1, y = V2)) +
  geom_point(aes(fill = `Region:Site`), shape = 21, size = 2, alpha = 0.75) +
  # scale_fill_manual(values = site_colors) +
  # facet_grid(Region ~ .) +
  coord_fixed(pcvar_c1[2]/pcvar_c1[1]) + 
  theme(legend.position = "right") +
  xlab(paste0( "PC1 (", pcvar_c1[1], "%)")) + 
  ylab(paste0( "PC2 (", pcvar_c1[2], "%)"))
# 
```
```{r}
pcs_c1_spp <- pcs_c1 %>% mutate(SubSpecies = ifelse(V2 < -0.05, "C", (ifelse(V1 < -0.05, "B", ifelse(V1 > 0, "A", "D"))))) 
```
```{r}
pdf(file="./outputs/figures/Evolution2024_pcangsd_pgrandis_region.pdf", width = 7.5, height = 5)
title <- expression(paste("East Pacific ", italic("P. grandis"), " population-level PCA (3,273,302 SNPs)"))
# no bone Pcas 
pcs_c1 %>% ggplot(aes(x = V1, y = V2)) +
  geom_point(aes(fill = `NGSAdmix K5 Species`), shape = 21, size = 2, alpha = 0.75, show.legend = T) +
  scale_fill_manual(values = c("darkorange4", "darkorange", "grey90")) +
  # facet_wrap(~ Region) +
  coord_fixed(pcvar_c1[2]/pcvar_c1[1]) + 
  xlab(paste0( "PC1 (", pcvar_c1[1], "%)")) + 
  ylab(paste0( "PC2 (", pcvar_c1[2], "%)")) + 
  theme(plot.title = element_text(size = 16, hjust = 0),
        axis.title = element_text(size = 14),
        axis.text = element_text(size = 12),
        legend.position = "right", 
        legend.spacing.y = unit(2, "mm"),
        legend.title = element_blank(),
        legend.text = element_text(size = 12)) +
  ggtitle(title)
dev.off()
```
```{r}
png(file="./outputs/figures/sicb/pcangsd_pgrandis_region.png", width = 8, height = 3.75, units = "in", res = 300)
title <- expression(paste(italic("P. grandis"), " population-level PCA (3,273,302 SNPs)"))
# no bone Pcas 
pcs_c1 %>% ggplot(aes(x = V1, y = V2)) +
  geom_point(aes(fill = `Region`), shape = 21, size = 2, alpha = 0.75, show.legend = T) +
  scale_fill_manual(values = region_colors) +
  # facet_wrap(~ Region) +
  coord_fixed(pcvar_c1[2]/pcvar_c1[1]) + 
  xlab(paste0( "PC1 (", pcvar_c1[1], "%)")) + 
  ylab(paste0( "PC2 (", pcvar_c1[2], "%)")) + 
  theme(plot.title = element_text(size = 16, hjust = 0.1),
        axis.title = element_text(size = 14),
        axis.text = element_text(size = 12),
        legend.position = "right", 
        legend.spacing.y = unit(2, "mm"),
        legend.title = element_blank(),
        legend.text = element_text(size = 12)) +
  ggtitle(title)
dev.off()
```

### Plot PCAs for *P. verrucosa* (3a + 3b + 3d) - no clones
```{r pca_c3}
C <- as.matrix(read.table("./outputs/angsd/pverrucosa_full_noclones.cov"))
e <- eigen(C)
```
```{r pca_test_plot}
plot(e$vectors[,1:2],xlab="PC1",ylab="PC2", main="individual allele frequency")
#
plot(e$vectors[,2:3],xlab="PC2",ylab="PC3", main="individual allele frequency")
# 
```
```{r bamfile_sample_order_check}
# reading list of bam files = order of samples in IBS matrix
bams <- read.table("/Users/mikeconnelly/computing/projects/etp_pocillopora_gskim/outputs/angsd/pverrucosa_full_noclones_bamfile.txt",header=F)[,1]
#
bam_ids <- sub("/scratch/nmnh_corals/connellym/projects/etp_pocillopora_gskim/outputs/alignments/", "", bams, perl=T)
bam_ids <- sub(".sorted.md.rg.bam", "", bam_ids, perl=T)
bam_ids <- sub("_S[[:digit:]]+$", "", bam_ids)
#
# bam_ids
```
```{r select_samples}
# aligning sample metadata with bams list, then check again
samples_pverrucosa <- samples[match(bam_ids, samples$`Sample ID`),] 
# do all the sample ids match the bam file ids in the correct order?
all(samples_pverrucosa$`Sample ID` == bam_ids) # TRUE
```
```{r combine_pca_metadata}
# combine pc coordinates with sample metadata
pcs <- as.data.frame(e$vectors)# %>% cbind(pverrucosa) %>% left_join(samples)
# store variable with % variance explained by each PC
pcvar <- round(e$values, 2)
#
```
```{r combine_pca_metadata}
# combine pc coordinates with sample metadata
pcs <- as.data.frame(e$vectors) %>% cbind(samples_pverrucosa)
# store variable with % variance explained by each PC
pcvar <- round(e$values, 2)
#
```
```{r pca_cluster_ggplot}
# test ggplot 
pcs %>% ggplot(aes(x = V1, y = V2)) +
  geom_point(aes(fill = `Region`), shape = 21, size = 2, alpha = 0.75) +
  scale_fill_manual(values = region_colors) +
  # facet_grid(Region ~ .) +
  coord_fixed(pcvar[2]/pcvar[1]) +
  xlab(paste0( "PC1 (", pcvar[1], "%)")) +
  ylab(paste0( "PC2 (", pcvar[2], "%)"))
# 
```

### Figure ###
```{r pverrucosa_region_pca}
pdf("./outputs/figures/pverrucosa_region_pca.pdf", width = 8, height = 3.75)
title <- expression(paste(italic("P. verrucosa"), " population-level PCA (42 samples, 13,224,661 SNPs)"))
#
pcs %>% ggplot(aes(x = V1, y = V2)) +
  geom_point(aes(fill = `Region`), shape = 21, size = 2, alpha = 0.75) +
  scale_fill_manual(values = region_colors) +
  # facet_grid(Region ~ .) +
  coord_fixed(pcvar[2]/pcvar[1]) + 
  theme(legend.position = "right") +
  xlab(paste0( "PC1 (", pcvar[1], "%)")) + 
  ylab(paste0( "PC2 (", pcvar[2], "%)")) +
  theme(plot.title = element_text(size = 16, hjust = 0.1),
        axis.title = element_text(size = 14),
        axis.text = element_text(size = 12),
        legend.position = "right", 
        legend.spacing.y = unit(2, "mm"),
        legend.title = element_blank(),
        legend.text = element_text(size = 12)) +
  ggtitle(title)
# 
dev.off()
```


## No-clones PCAngsd admixture (unlinked SNPs)
```{r}
#open R
##Requires previous installation of the library RcppCNPy

library("RcppCNPy") # Numpy library for R

# pop<-read.table("Demo/Data/Demo1pop.info",as.is=T)

# q <- npyLoad("./outputs/angsd/final_noclones.admix.Q.npy") 
q <- npyLoad("./outputs/angsd/final_noclones_noLD.admix.Q.npy") 


q_pop <- cbind(samples_noclones, q)

barplot(t(q),col=2:10,space=0,border=NA,xlab="Individuals",ylab="Admixture proportions")

## close R
```
```{r}
q_pop_tidy <- q_pop %>%
  mutate("Sample ID" = fct_reorder(`Sample ID`, desc(Species))) %>% 
  pivot_longer(cols = `1`:`5`, names_to = "pop") %>%
  arrange(Region, mtorf_type)
#
#
q_pop_tidy_plot <- q_pop_tidy %>%
  ggplot(aes(x = `Sample ID`, y = value)) +
  geom_col(aes(fill = `pop`), position = "stack", show.legend = F) +
  # scale_fill_manual(values = admix_pop_colors) +
    facet_grid(~Region, scales = "free", space = "free") + 
  xlab("Individuals") +
  # ylab(paste0("K=",i)) +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        legend.text = element_text(size = 2))

print(q_pop_tidy_plot)
```

```{r}
`%notin%` <- Negate(`%in%`)
q_pop_rep_samples <- q_pop_tidy %>% filter(value >= 0.85) 

q_pop_tidy %>% 
  filter(`Sample ID` %in% q_pop_rep_samples$`Sample ID`) %>% 
      mutate("Sample ID" = fct_reorder2(`Sample ID`, desc(pop), pop),
         "Region" = factor(`Region`, levels = region_order_longitude, ordered = T)) %>% 
  ggplot(aes(x = `Sample ID`, y = value)) +
  geom_col(aes(fill = `pop`), position = "stack", show.legend = F) +
  scale_fill_manual(values = admix_pop_colors) +
    facet_grid(~Region, scales = "free", space = "free") + 
  xlab("Individuals") +
  # ylab(paste0("K=",i)) +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        legend.text = element_text(size = 2))

q_pop_tidy %>% 
  filter(`Sample ID` %notin% q_pop_rep_samples$`Sample ID`) %>% 
      mutate("Sample ID" = fct_reorder2(`Sample ID`, desc(pop), pop),
         "Region" = factor(`Region`, levels = region_order_longitude, ordered = T)) %>% 
  ggplot(aes(x = `Sample ID`, y = value)) +
  geom_col(aes(fill = `pop`), position = "stack", show.legend = F) +
  scale_fill_manual(values = admix_pop_colors) +
    facet_grid(~Region, scales = "free", space = "free") + 
  xlab("Individuals") +
  # ylab(paste0("K=",i)) +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        legend.text = element_text(size = 2))
```