---
title: "NGSAdmix"
author: "Mike Connelly"
date: "2023-07-19"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, cache = TRUE)
options(stringsAsFactors = FALSE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```
```{r packages}
library("tidyverse")
library("RColorBrewer")
library("patchwork")
# 
source("./R/pocillopora_etp_gksim_functions.R")
#
# code adapted from: https://baylab.github.io/MarineGenomics/week-9-population-structure-using-ngsadmix.html and https://github.com/alexkrohn/AmargosaVoleTutorials/blob/master/ngsAdmix_tutorial.md
```

## Import sample metadata
```{r import_sample_metadata}
source("./R/sample_import.R")
samples <- all_samples
# NOTE: Remove clones from the dataset using IBS analysis before running NGSadmix
```

## No-clones samples (w/ linked SNPs)
```{r choose_best_K}
# Choose best K
logs <- read_delim("./outputs/ngsadmix/final_noclones/final_noclones_ngsadmix_likeKlogs.txt", trim_ws = TRUE, col_names = c("K", "value"), col_types = cols("K" = col_character(), "value" = col_double()))

logKdf <- logs %>% group_by(K) %>% 
  summarise(mK = (mean(abs(value), na.rm = T))/sd(abs(value), na.rm = T)) %>%  
  arrange(mK)

Kvec <- c("K1", "K2", "K3", "K4", "K5", "K6", "K7", "K8", "K9", "K10", "K11", "K12")
# Kvec <- seq(1,12)

logKdf$K <- factor(logKdf$K, levels = Kvec, ordered = T)

bestK <- as.numeric(gsub("K", "", logKdf$K[1]))

bestKplot <- logKdf %>% ggplot(aes(K, log(mK))) + 
  geom_vline(xintercept = bestK, color = "grey50") +
  geom_point()

print(bestKplot)
```
```{r save_bestK_plot}
pdf("./outputs/figures/ngsadmix_bestK_plot.pdf", width = 5.5, height = 3.5)
print(bestKplot)
dev.off()
```
```{r bam_list}
# Obtain sample info
# read in the population information file located in the bam.filelist
bams <- read.table("/Users/mikeconnelly/computing/projects/etp_pocillopora_gskim/outputs/angsd/final_noclones_bamfile.txt",header=F)[,1]
bam_ids <- sub("/scratch/nmnh_corals/connellym/projects/etp_pocillopora_gskim/outputs/alignments/", "", bams, perl=T)
bam_ids <- sub(".sorted.md.rg.bam", "", bam_ids, perl=T)
bam_ids <- sub("_S[[:digit:]]+$", "", bam_ids)
#
# bam_ids
pop <- data_frame("Sample ID" = bam_ids) %>% left_join(samples, by = "Sample ID")
```

### Plot ancestry proportions per K
```{r import_qopt_files}
# make combination of plots for multiple values of K
# for example, K from 1 - 10
maxK <- 10
popcolors <- brewer.pal(maxK, "Set1")
qopt_list <- list()
for (i in 1:maxK){
  print(i)
  kcols <- c()
  # internal loop to create column header names
  for (j in c(1:i)){
  kcols <- append(kcols, paste0("pop",j))
  }
  # print(kcols)
  qopt_list[[i]] <- read_table(paste0("./outputs/ngsadmix/final_noclones/final_noclones_K",i,"_1.qopt"), col_names = kcols)
}
qopt_list
```
```{r produce_plots}
qpop_list <- list()
plot_list <- list()
for (i in 1:maxK){
q <- qopt_list[[i]]
q_pop <- cbind(pop, q)
  ## ggplot attempt
q_pop_tidy <- q_pop %>%
  mutate("Sample ID" = fct_reorder(`Sample ID`, desc(Species))) %>% 
  pivot_longer(cols = contains("pop"), names_to = "pop") %>%
  arrange(Region, mtorf_type)
#
#
q_pop_tidy_plot <- q_pop_tidy %>%
  ggplot(aes(x = `Sample ID`, y = value)) +
  geom_col(aes(fill = `pop`), position = "stack", show.legend = F) +
  scale_fill_manual(values = popcolors) +
    facet_grid(~Region, scales = "free", space = "free") + 
  xlab("Individuals") +
  ylab(paste0("K=",i)) +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        legend.text = element_text(size = 2))
#
# print(q_pop_tidy_plot)
qpop_list[[i]] <- q_pop
plot_list[[i]] <- q_pop_tidy_plot
}
```
```{r print_plots}
pdf("./outputs/figures/ngsadmix_final_noclones_K1-6.pdf", width = 8.5, height = 8.5)
# print(plot_list[[1]])
title <- expression(paste("NGSadmix results for K=2-9 on East Pacific ", italic("Pocillopora"), " (5,314,313 SNPs)"))
plot_list[[2]] / plot_list[[3]] / plot_list[[4]] / plot_list[[5]] / plot_list [[6]] / plot_list[[7]] / plot_list[[8]] / plot_list[[9]] +
dev.off()
```
### Plot best K ancestry proportions
```{r print_plot_bestK}
plot_list[[bestK]]
```
```{r save_plot_bestK}
pdf("./outputs/figures/ngsadmix_final_noclones_K5.pdf", width = 8.5, height = 4.5)
title <- expression(paste("NGSadmix results for best K on East Pacific ", italic("Pocillopora"), " (5,314,313 SNPs)"))

plot_list[[bestK]] + ggtitle(title)
dev.off()
```
### Explore ancestry in best K results
```{r}
# explore assigned ancestry proportions in best K
qpop_best <- qpop_list[[bestK]]
#
q_pop_besttidy <- qpop_best %>%
  mutate("Sample ID" = fct_reorder(`Sample ID`, desc(Species))) %>% 
  pivot_longer(cols = contains("pop"), names_to = "pop") %>%
  arrange(Region, mtorf_type)
#
q_pop_besttidy %>% 
  filter(value > 0.85) %>% 
  group_by(pop) %>%
  count()

# 209 non-admixed individuals
# pop1 = Offshore P. grandis  (31)
# pop2 = P. verrucosa 3a (33)
# pop3 = P. meandrina (10)
# pop4 = Continental P. grandis (131)
# pop5 = P. effusa (4)
```
```{r}
qpop_best %>% filter(pop5 > 0.85) %>% select(`Sample ID`, `Region:Site`, mtorf_type, Species) %>% View()
```
```{r}
# identify representative individuals
qpop_rep_samples <- q_pop_besttidy %>% 
  filter(value > 0.85)
  
# identify admixed individuals
q_pop_besttidy %>% 
  filter(value > 0.10, value < 0.85) %>% 
  filter(pop == "pop5") %>% # ggplot(aes(value)) + geom_histogram(binwidth = 0.01)
  group_by(pop) %>% # count()
  View()

# 5 samples from Coco are admixed from pop2 and pop5!
```
```{r}
pdf("./outputs/figures/NGSAdmix_K5_results_plots.pdf", width = 11.5, height = 3.5)
q_pop_besttidy %>%
  # filter(`Sample ID` %in% qpop_rep_samples$`Sample ID`) %>% 
  ggplot(aes(x = `Sample ID`, y = value)) +
  geom_col(aes(fill = `pop`), position = "stack", show.legend = T) +
  scale_fill_manual(values = popcolors) +
    facet_grid(~Region, scales = "free", space = "free") +
  xlab("Individuals") +
  # ylab(paste0("K=",i)) +
  theme(axis.text.x = element_blank(),
        strip.text.x = element_text(size = 6),
        axis.title.x = element_blank(),
        legend.text = element_text(size = 8)) +
    ggtitle("All no-clones individuals")

q_pop_besttidy %>%
  filter(`Sample ID` %in% qpop_rep_samples$`Sample ID`) %>% 
  ggplot(aes(x = `Sample ID`, y = value)) +
  geom_col(aes(fill = `pop`), position = "stack", show.legend = T) +
  scale_fill_manual(values = popcolors) +
    facet_grid(~Region, scales = "free", space = "free") +
  xlab("Individuals") +
  # ylab(paste0("K=",i)) +
  theme(axis.text.x = element_blank(),
        strip.text.x = element_text(size = 6),
        axis.title.x = element_blank(),
        legend.text = element_text(size = 8)) +
    ggtitle("Representative individuals (q > 0.85)")

`%notin%` <- Negate(`%in%`)

q_pop_besttidy %>%
  filter(`Sample ID` %notin% qpop_rep_samples$`Sample ID`) %>% 
  ggplot(aes(x = `Sample ID`, y = value)) +
  geom_col(aes(fill = `pop`), position = "stack", show.legend = T) +
  scale_fill_manual(values = popcolors) +
    facet_grid(~Region, scales = "free", space = "free") +
  xlab("Individuals") +
  # ylab(paste0("K=",i)) +
  theme(axis.text.x = element_blank(),
        strip.text.x = element_text(size = 6),
        axis.title.x = element_blank(),
        legend.text = element_text(size = 8)) + 
  ggtitle("Admixed individuals")

dev.off()
```

## No-clones samples (unlinked SNPs after LD pruning)
```{r choose_best_K}
# Choose best K
logs <- read_delim("./outputs/ngsadmix/final_noclones/final_noclones_noLD_ngsadmix_likeKlogs.txt", trim_ws = TRUE, col_names = c("K", "value"), col_types = cols("K" = col_character(), "value" = col_double()))
# *P. grandis*
logs <- read_delim("./outputs/ngsadmix/final_noclones/final_noclones_pgra_ngsadmix_likeKlogs.txt", trim_ws = TRUE, col_names = c("K", "value"), col_types = cols("K" = col_character(), "value" = col_double()))


logKdf <- logs %>% group_by(K) %>% 
  summarise(mK = (mean(abs(value), na.rm = T))/sd(abs(value), na.rm = T)) %>%  
  arrange(mK)

Kvec <- c("K1", "K2", "K3", "K4", "K5", "K6", "K7", "K8", "K9", "K10", "K11", "K12")
Kvec <- seq(1,12)

logKdf$K <- factor(logKdf$K, levels = Kvec, ordered = T)

bestK <- as.numeric(gsub("K", "", logKdf$K[1]))

bestKplot <- logKdf %>% ggplot(aes(K, log(mK))) + 
  geom_vline(xintercept = bestK, color = "grey50") +
  geom_point()

print(bestKplot)
# K=5 is the "best" K, but K=6 is very close and separates the Isla del Coco samples 
```
### Figure S5
```{r figS5_bestK_plot}
pdf("./outputs/figures/FigS5_ngsadmix_noLD_bestK_plot.pdf", width = 6.5, height = 3.5)
print(bestKplot)
dev.off()
```
```{r bam_list}
# Obtain sample info
# read in the population information file located in the bam.filelist
bams <- read.table("./outputs/angsd/final_noclones_pgra_bamfile.txt",header=F)[,1]
#
bam_ids <- sub("/scratch/nmnh_corals/connellym/projects/etp_pocillopora_gskim/outputs/alignments/pgra_himb/", "", bams, perl=T)
bam_ids <- sub(".sorted.md.rg.bam", "", bam_ids, perl=T)
bam_ids <- sub("_S[[:digit:]]+$", "", bam_ids)
#
# bam_ids
pop <- data_frame("Sample ID" = bam_ids) %>% left_join(samples, by = "Sample ID")
```
### Plot ancestry proportions per K
```{r import_qopt_files}
# make combination of plots for multiple values of K
# for example, K from 1 - 10
maxK <- 12
# define color set
popcolors_maxK <- brewer.pal(maxK, "Set1")

qopt_list <- list()
for (i in 1:maxK){
  print(i)
  kcols <- c()
  # internal loop to create column header names
  for (j in c(1:i)){
  kcols <- append(kcols, paste0("pop",j))
  }
  # print(kcols)
  qopt_list[[i]] <- read_table(paste0("./outputs/ngsadmix/final_noclones/final_noclones_noLD_pgra_himb_K",i,"_1.qopt"), col_names = kcols)
}
# qopt_list
```
```{r produce_plots}
qpop_list <- list()
plot_list <- list()
for (i in 1:maxK){
q <- qopt_list[[i]]
q_pop <- cbind(pop, q)
  ## ggplot attempt
q_pop_tidy <- q_pop %>%
  mutate("Sample ID" = fct_reorder(`Sample ID`, desc(Species)),
         "Region" = factor(`Region`, levels = region_order_longitude, ordered = T)) %>% 
  pivot_longer(cols = contains("pop"), names_to = "pop") %>%
  arrange(Region, mtorf_type)
#
# define color set
popcolors <- popcolors_maxK[1:i]
#
q_pop_tidy_plot <- q_pop_tidy %>%
  ggplot(aes(x = `Sample ID`, y = value)) +
  geom_col(aes(fill = `pop`), position = "stack", show.legend = F) +
  scale_fill_manual(values = popcolors) +
    facet_grid(~Region, scales = "free", space = "free") + 
  xlab("Individuals") +
  ylab(paste0("K=",i)) +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        legend.text = element_text(size = 2))
#
# print(q_pop_tidy_plot)
qpop_list[[i]] <- q_pop
plot_list[[i]] <- q_pop_tidy_plot
}
```
```{r print_plots_K1-9}
pdf("./outputs/figures/ngsadmix_final_noclones_noLD_K1-9.pdf", width = 8.5, height = 8.5)
# print(plot_list[[1]])
title <- expression(paste("NGSadmix results for K=2-10 on East Pacific ", italic("Pocillopora"), " (1,310,435 SNPs)"))
plot_list[[2]] / plot_list[[3]] / plot_list[[4]] / plot_list[[5]] / plot_list [[6]] / plot_list[[7]] / plot_list[[8]] / plot_list[[9]]
dev.off()
```

### Plot best K ancestry proportions
```{r print_plot_bestK}
plot_list[[5]]
```
```{r save_plot_bestK}
pdf("./outputs/figures/ngsadmix_final_noclones_noLD_K5.pdf", width = 8.5, height = 4.5)
title <- expression(paste("NGSadmix results for best K on East Pacific ", italic("Pocillopora"), " (1,310,435 SNPs)"))

plot_list[[bestK]] + ggtitle(title)
dev.off()
```

### Explore ancestry in best K results (K=5)
```{r bestK_ancestry_assignments}
# explore assigned ancestry proportions in best K
qpop_best <- qpop_list[[bestK]]
# Using best K admixture results, identify representative samples from clusters (q > 0.85)
representative <- function(x) {x >= 0.85}

# create new metadata columns to identify representative vs. admixed samples and assign population labels
samples_noclones_ngsadmix <- qpop_best %>% mutate(`NGSADMIX_REP_K5` = if_any(contains("pop"), representative),
                     `NGSAdmix K5 Species` = ifelse(representative(pop1), "P. meandrina", 
                                           ifelse(representative(pop2), "P. effusa",
                                                  ifelse(representative(pop3), "P. grandis - Offshore",
                                                         ifelse(representative(pop4), "P. verrucosa 3a",
                                                                ifelse(representative(pop5), "P. grandis - Continent", "Unassigned")))))) %>%
  select(`Sample ID`, `NGSADMIX_REP_K5`, `NGSAdmix K5 Species`)
# next, join this to no-clones sample data for use in maps, PCA plots, etc.
write_csv(samples_noclones_ngsadmix, "./data/samples_noclones_ngsadmix_assignments_k5.csv")
```
```{r bestK_representative_samples}
qpop_rep_samples <- qpop_best %>% filter(if_any(contains("pop"), representative))
#
rep_samples <- qpop_rep_samples$`Sample ID`
#
q_pop_besttidy <- qpop_best %>%
  mutate("Sample ID" = fct_reorder(`Sample ID`, desc(Species))) %>% 
  pivot_longer(cols = contains("pop"), names_to = "pop") %>%
  arrange(Region, mtorf_type)
#
q_pop_besttidy %>% 
  filter(value >= 0.85) %>% 
  group_by(pop) %>%
  count()

# 210 non-admixed individuals
# pop1 = P. meandrina (10)
# pop2 = P. effusa (4)
# pop3 = Offshore P. grandis  (33)
# pop4 = P. verrucosa 3a (33) (P. lacera?)
# pop5 = Continental P. grandis (130) (P. capitata?) 
```
```{r bestK_admixed_samples}
# Using best K admixture results, identify admixed/unassigned samples from clusters (all q < 0.85)
admixed <- function(x) {x <= 0.85} 
qpop_admixed_samples <- qpop_best %>% filter(if_all(contains("pop"), admixed))
#
admixed_samples <- qpop_admixed_samples$`Sample ID`
```
```{r compare_representative_admixed_samples_plots}
pdf("./outputs/figures/NGSAdmix_noLD_K5_results_plots.pdf", width = 11.5, height = 3.5)

q_pop_besttidy %>%
 left_join(samples_noclones_ngsadmix, by = "Sample ID") %>% 
 mutate("Region" = factor(`Region`, levels = region_order_longitude, ordered = T),
        "NGSAdmix Species" = factor(`NGSAdmix Species`, levels = ngsadmix_spp, ordered = T),
        "pop" = factor(`pop`, levels = c("pop1", "pop4", "pop3", "pop2", "pop5"), ordered = T),
        "Sample ID" = fct_reorder2(`Sample ID`, `NGSAdmix Species`, desc(`NGSAdmix Species`))) %>% 
  ggplot(aes(x = `Sample ID`, y = value)) +
  geom_col(aes(fill = `pop`), position = "stack", show.legend = T) +
  scale_fill_manual(values = unname(ngsadmix_pop_colors)) +
    facet_grid(~Region, scales = "free", space = "free") +
  xlab("Individuals") +
  # ylab(paste0("K=",i)) +
  theme(axis.text.x = element_blank(),
        strip.text.x = element_text(size = 6),
        axis.title.x = element_blank(),
        legend.text = element_text(size = 8)) +
    ggtitle("All no-clones individuals") 

q_pop_besttidy %>%
  filter(`Sample ID` %in% rep_samples) %>% 
    mutate("Sample ID" = fct_reorder2(`Sample ID`, desc(pop), pop),
         "Region" = factor(`Region`, levels = region_order_longitude, ordered = T)) %>% 
  ggplot(aes(x = `Sample ID`, y = value)) +
  geom_col(aes(fill = `pop`), position = "stack", show.legend = T) +
  scale_fill_manual(values = popcolors) +
    facet_grid(~Region, scales = "free") +
  xlab("Individuals") +
  # ylab(paste0("K=",i)) +
  theme(axis.text.x = element_blank(),
        strip.text.x = element_text(size = 6),
        axis.title.x = element_blank(),
        legend.text = element_text(size = 8)) +
    ggtitle("Representative individuals (q > 0.85)")

`%notin%` <- Negate(`%in%`)

q_pop_besttidy %>%
  filter(`Sample ID` %notin% rep_samples) %>% 
  ggplot(aes(x = `Sample ID`, y = value)) +
  geom_col(aes(fill = `pop`), position = "stack", show.legend = T) +
  scale_fill_manual(values = popcolors) +
    facet_grid(~Region, scales = "free", space = "free") +
  xlab("Individuals") +
  # ylab(paste0("K=",i)) +
  theme(axis.text.x = element_blank(),
        strip.text.x = element_text(size = 6),
        axis.title.x = element_blank(),
        legend.text = element_text(size = 8)) + 
  ggtitle("Admixed individuals")

dev.off()
```

### Explore ancestry in best K results (K=6)
```{r K6_ancestry_assignments}
# explore assigned ancestry proportions in best K
qpop_best <- qpop_list[[6]]
# Using best K admixture results, identify representative samples from clusters (q > 0.85)
representative <- function(x) {x >= 0.85}

# create new metadata columns to identify representative vs. admixed samples and assign population labels
samples_noclones_ngsadmix <- qpop_best %>% mutate(`NGSADMIX_REP_K6` = if_any(contains("pop"), representative),
                     `NGSAdmix K6 Species` = ifelse(representative(pop1), "P. grandis - Continent", 
                                           ifelse(representative(pop2), "P. meandrina",
                                                  ifelse(representative(pop3), "P. grandis - Offshore",
                                                         ifelse(representative(pop4), "P. effusa",
                                                                ifelse(representative(pop5), "P. verrucosa 3b",
                                                                       ifelse(representative(pop6), "P. verrucosa 3a", "Unassigned"))))))) %>%
  dplyr::select(`Sample ID`, `NGSADMIX_REP_K6`, `NGSAdmix K6 Species`)
# next, join this to no-clones sample data for use in maps, PCA plots, etc.
write_csv(samples_noclones_ngsadmix, "./data/samples_noclones_ngsadmix_assignments_k6.csv")
```

### Figure 1B

```{r print_plots_K5-6}
# Plot admixture results for K=5 and K=6 since they are essentially equally likely
# Ensure colors are consistent between plots to highlight the slight differences (requires manual edits to color vectors)
# define color sets
# popcolors_k5 <- popcolors_maxK[1:5]
popcolors_k5 <- unname(ngsadmix_pop_colors)[c(2,1,3,5,4)]
# popcolors_k6 <- popcolors_maxK[c(2,4,3,1,6,5)]
popcolors_k6 <- unname(ngsadmix_pop_colors_k6[c(4,2,3,1,6,5)])

plot_k5 <- plot_list[[5]] + scale_fill_manual(values = popcolors_k5) + theme(strip.text = element_text(size = 4), axis.text = element_text(size = 6))
plot_k5
plot_k6 <- plot_list[[6]] + scale_fill_manual(values = popcolors_k6) + theme(strip.text = element_text(size = 4), axis.text = element_text(size = 6))
# plot_k6

# pdf("./outputs/figures/ngsadmix_final_noclones_noLD_K5-6.pdf", width = 8.5, height = 3.5)
# print(plot_list[[1]])

title <- expression(paste("NGSadmix results for K=5-6 on East Pacific ", italic("Pocillopora"), " (1,310,435 SNPs)"))

plot_k5 / plot_k6

# dev.off()
```

```{r region_abbreviations}
regions_abbv <- list(
"Clipperton" = "CLIP",
"Galapagos Northern" = "GPN",
"Galapagos Central" = "GPC",
"Coco" = "COCO",
"Bahia Culebra" = "BCU",
"Golfo Dulce" = "GDU",
"Chiriqui" = "CHIQ",
"Panama" = "PAN",
"Gorgona" = "GOR"
)
abbv_labeller <- function(variable, value){
  return(regions_abbv[value])
}
```
```{r easy_ancestry_assignments}
# tweak ancestry assignments to eliminate "Unassigned" samples - helps ease gradients in barplots and doesn't separate samples with admixture
qpop_best <- qpop_list[[bestK]]
# Using best K admixture results, identify representative samples from clusters (q > 0.85)
representative <- function(x) {x >= 0.5}

# create new metadata columns to identify representative vs. admixed samples and assign population labels
samples_k5_noadmix_assignments <- qpop_best %>% mutate(`NGSADMIX_REP_K5` = if_any(contains("pop"), representative),
                     `NGSAdmix K5 Species` = ifelse(representative(pop1), "P. meandrina", 
                                           ifelse(representative(pop2), "P. effusa",
                                                  ifelse(representative(pop3), "P. grandis - Offshore",
                                                         ifelse(representative(pop4), "P. verrucosa 3a",
                                                                ifelse(representative(pop5), "P. grandis - Continent",
                                                                       "Unassigned")))))) %>%
  dplyr::select(`Sample ID`, `NGSADMIX_REP_K5`, `NGSAdmix K5 Species`)

ngsadmix_spp_code <- c("P. effusa" = "A",
                  "P. meandrina" = "B",
                  "P. grandis - Offshore" = "C",
                  "P. grandis - Continent" = "D",
                  "P. verrucosa 3a" = "E")
```


```{r barplot_k5}
popcolors_k5 <- unname(ngsadmix_pop_colors_k6[c(2,1,3,5,4)])
names(popcolors_k5) <- c("pop1", "pop2", "pop3", "pop4", "pop5")

q_k5 <- qopt_list[[5]]
q_pop_k5 <- cbind(pop, q_k5) %>%
  # left_join(samples_k5_assignments, by = "Sample ID") %>%
  left_join(samples_k5_noadmix_assignments, by = "Sample ID") %>% #samples_k6_assignments
  # ensure correct sample order for plotting and region order for facets
  mutate("Sample ID" = factor(`Sample ID`, levels = .$`Sample ID`, ordered = T))

q_pop_k5$`NGSAdmix K5 Spp Label` <- ngsadmix_spp_code[q_pop_k5$`NGSAdmix K5 Species`]
         
 q_pop_k5 <- q_pop_k5 %>% mutate(
            "Sample ID" = fct_reorder2(`Sample ID`, `pop1`, `pop3`), 
            "Sample ID" = fct_reorder(`Sample ID`, `pop5`),
            "Sample ID" = fct_reorder2(`Sample ID`, `pop3`, desc(`pop4`)),
            "Sample ID" = fct_reorder(`Sample ID`, `NGSAdmix K5 Spp Label`),
         # 
         "Region" = factor(`Region`, levels = region_order_longitude, ordered = T)) %>% 
  arrange(`Sample ID`)

# save Sample ID order for reordering factors in K = 6 plot
sample_order <- q_pop_k5$`Sample ID`
  
q_pop_tidy_k5 <- q_pop_k5 %>%
  pivot_longer(cols = contains("pop"), names_to = "pop") %>%
  # ensure correct order of populations for stacking barplots
  mutate("pop" = factor(`pop`, levels = c("pop2", "pop1", "pop3", "pop5", "pop4"), ordered = T))


q_pop_tidy_plot_k5 <- q_pop_tidy_k5 %>%
  ggplot(aes(x = `Sample ID`, y = value)) +
  geom_col(aes(fill = `pop`), position = "stack", show.legend = F) +
  scale_fill_manual(values = popcolors_k5) +
  facet_grid(~Region, scales = "free", space = "free", labeller = abbv_labeller) + 
  ylab("K = 5") +
  scale_y_continuous(breaks = c(0, 0.5, 1), expand = expansion(mult = c(0, 0))) + 
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 12),
        strip.text = element_text(size = 6, margin = margin(0.5, 0, 0.5, 0, unit = "mm")),
        strip.background = element_rect(fill = "white"),
        axis.text.y = element_text(size = 6),
        # axis.text.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.ticks.y = element_line(size = 0.25),
        axis.ticks.length.y = unit(2, "pt"),
        panel.grid = element_blank(),
        panel.spacing.x = unit(0.5, "mm"),
        plot.margin = margin(2, 2, 2, 2, "mm"))

q_pop_tidy_plot_k5
```
```{r barplot_k6}
popcolors_k6 <- unname(ngsadmix_pop_colors_k6[c(4,2,3,1,6,5)])
names(popcolors_k6) <- c("pop1", "pop2", "pop3", "pop4", "pop5", "pop6")

q_k6 <- qopt_list[[6]]
q_pop_k6 <- cbind(pop, q_k6) %>%
  left_join(samples_k5_assignments, by = "Sample ID") %>%
  left_join(samples_k6_assignments, by = "Sample ID") %>% 
  # ensure correct sample order for plotting and region order for facets
  mutate("Sample ID" = factor(`Sample ID`, levels = sample_order, ordered = T),
         "Region" = factor(`Region`, levels = region_order_longitude, ordered = T)) %>% 
  arrange(`Sample ID`)
  
q_pop_tidy_k6 <- q_pop_k6 %>%
  pivot_longer(cols = contains("pop"), names_to = "pop") %>% 
  # ensure correct order of populations for stacking barplots
  mutate("pop" = factor(`pop`, levels = c("pop4", "pop2", "pop3", "pop1", "pop6", "pop5"), ordered = T))

q_pop_tidy_plot_k6 <- q_pop_tidy_k6 %>%
  ggplot(aes(x = `Sample ID`, y = value)) +
  geom_col(aes(fill = `pop`), position = "stack", show.legend = F) +
  scale_fill_manual(values = popcolors_k6) +
  facet_grid(~Region, scales = "free", space = "free", labeller = abbv_labeller) + 
  ylab("K = 6") +
  scale_y_continuous(breaks = c(0, 0.5, 1), expand = expansion(mult = c(0, 0))) + 
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 12),
        strip.text = element_text(size = 6, margin = margin(0.5, 0, 0.5, 0, unit = "mm")),
        strip.background = element_rect(fill = "white"),
        # strip.text = element_blank(),
        axis.text.y = element_text(size = 6),
        # axis.text.y = element_blank(),
        axis.ticks.x = element_blank(), 
        axis.ticks.y = element_line(size = 0.25), 
        axis.ticks.length.y = unit(2, "pt"),
        panel.grid = element_blank(),
        panel.spacing.x = unit(0.5, "mm"),
        plot.margin = margin(0.5, 0, 0, 0, "mm"))

q_pop_tidy_plot_k6
```

```{r fig1B_ngsadmix_barplot_k5-6}
pdf("./outputs/figures/Fig1B_ngsadmix_K5_barplot.pdf", width = 6.65, height = 1.5)
q_pop_tidy_plot_k5 
dev.off()
```

```{r fig1B_ngsadmix_barplot_k5-6}

# pdf("./outputs/figures/Fig1B_ngsadmix_K5_ancestrybarplot.pdf", width = 12.5, height = 3.5)
ngsadmix_K5_qopt_barplot <- q_pop_besttidy %>%
 left_join(samples_noclones_ngsadmix, by = "Sample ID") %>% 
 # filter(`Sample ID` %in% rep_samples) %>% 
 mutate("Region" = factor(`Region`, levels = region_order_longitude, ordered = T),
        "NGSAdmix Species" = factor(`NGSAdmix Species`, levels = ngsadmix_spp, ordered = T),
        "pop" = factor(`pop`, levels = c("pop1", "pop4", "pop3", "pop2", "pop5", "pop6"), ordered = T),
        "Sample ID" = fct_reorder2(`Sample ID`, desc(`NGSAdmix Species`), desc(`NGSAdmix Species`))) %>% 
  ggplot(aes(x = `Sample ID`, y = value)) +
  geom_col(aes(fill = `pop`), position = "stack", show.legend = F) +
  scale_fill_manual(values = unname(ngsadmix_pop_colors_k6)) +
    facet_grid(. ~ Region, scales = "free", space = "free") + # + `NGSAdmix Species` NGSADMIX_REPRESENTATIVE
  xlab("Individuals") +
  ylab("NGSAdmix Proportion Assigned Ancestry") +
  theme(axis.text.x = element_blank(),
        strip.text.x = element_text(size = 5),
        axis.title.x = element_blank(),
        legend.text = element_text(size = 8)) 
print(ngsadmix_K5_qopt_barplot)
# dev.off()
```

### Output lists of samples 
```{r sample_lists}
# write lists of representative samples for use in pairwise Fst estimation, phylogenetic tree construction, SNAPP BFD*, etc.
peffusa <- qpop_rep_samples %>% filter(pop1 > 0.85) %>% select(`Sample ID`)
pgrandiscontinent <- qpop_rep_samples %>% filter(pop2 > 0.85) %>% select(`Sample ID`)
pgrandisoffshore <- qpop_rep_samples %>% filter(pop3 > 0.85) %>% select(`Sample ID`)
pmeandrina <- qpop_rep_samples %>% filter(pop4 > 0.85) %>% select(`Sample ID`)
pverrucosa3a <- qpop_rep_samples %>% filter(pop5 > 0.85) %>% select(`Sample ID`)
#
write_lines(pgrandiscontinent$`Sample ID`, "~/Desktop/pgrandiscontinent")
write_lines(pgrandisoffshore$`Sample ID`, "~/Desktop/pgrandisoffshore")
write_lines(pmeandrina$`Sample ID`, "~/Desktop/pmeandrina")
write_lines(pverrucosa3a$`Sample ID`, "~/Desktop/pverrucosa3a")
```