---
title: "PHYLUCE_phylogenetic_analysis"
author: "Mike Connelly"
date: "2023-05-24"
output: html_document
---
## Setup and packages
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, cache = TRUE)
options(stringsAsFactors = FALSE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```
```{r packages}
library("tidyverse")
library("treeio")
library("ggtree")
library("phytools")
# 
source("./R/pocillopora_etp_gksim_functions.R")
```

## Import sample metadata
```{r import_sample_metadata}
source("./R/sample_import.R")
samples <- all_samples
```
```{r select_samples}
samples <- filter(all_samples, ANALYSIS ==T)
# samples <- filter(all_samples, `PANAMA_MANUSCRIPT` == T)
```
```{r outgroup_metadata}
# outgroups <- read_csv("./data/outgroups.csv")
# samples_out <- rbind(samples, outgroups)
```

## Phylogenetic trees from PHYLUCE UCE alignment

### IQ-TREE
```{r}
tree <- read.newick("./outputs/iqtree/iqtree_phyluce_all_noclones.treefile")
# midpoint root
tree <- midpoint.root(tree)
# set root-edge
tree$root.edge <- 1
gg_tree(tree)
```
```{r modify_tip_labels}
# remove contigs suffix 
tips <- gsub("_S.*_contigs$", "", tree$tip.label)
tree$tip.label <- tips
# remove biorepository codes 
tips_short <- gsub("_AP.*", "", tips)
```

```{r basic_ggtree_viz}
pdf("./outputs/figures/iqtree_02.pdf", width = 8, height = 8)
tree_title <- expression(paste("IQ-TREE Maximum Likelihood Tree of ", italic("Pocillopora"), " samples"))
#
tree %>% ggtree(., right = TRUE) %<+% samples + 
  geom_tippoint(aes(fill = `mtorf_type`), shape = 21, alpha = 0.8, stroke = 0.5, color = "black", hjust = 0.3, show.legend = TRUE) +
  geom_tiplab(size = 2, hjust = -0.05) +
  ggplot2::xlim(-0.1, 0.33) +
  geom_treescale(width = 0.04, x = -0.05, y = 1) +
  scale_fill_manual(values = mtorf_colors, name = "mtORF Type") +
   theme(legend.position = "right",
        legend.margin = margin(0,0,0,0, unit = "mm"),
        plot.margin = margin(5,5,5,5, unit = "mm")) +
  ggtitle(tree_title)
dev.off()
```
```{r annotated_ggtree_viz}
# clades - node 103: P. effusa, node 108: P. grandis, node 180: P. meandrina, node 183: P. verrucosa-like
pdf("./outputs/figures/iqtree_02.pdf", width = 8.5, height = 11.5)
#
tree %>% ggtree(., right = TRUE) %<+% samples + 
  geom_tippoint(aes(fill = `mtorf_type`), shape = 21, alpha = 0.8, stroke = 0.5, color = "black", hjust = 0.3, show.legend = TRUE) +
  geom_tiplab(size = 3, hjust = -0.05) +
  # add bootstrap support values
  geom_nodelab(size = 2, hjust = 1.25, vjust = -0.25) +
  # add clade annotations
  geom_cladelab(node = 103, label = "P. effusa (Clipperton)", angle = 0, 
                  fontsize = 3, offset = 0.15) +
  geom_cladelab(node = 108, label = "P. grandis-like", angle = 0, 
                  fontsize = 3, offset = 0.15) +
  geom_cladelab(node = 180, label = "P. meandrina-like", angle = 0, 
                  fontsize = 3, offset = 0.15) +
  geom_cladelab(node = 183, label = "P. verrucosa-like", angle = 0, 
                  fontsize = 3, offset = 0.15) +
  ggplot2::xlim(-0.05, 0.5) +
  geom_treescale(width = 0.05, x = 0) +
  scale_fill_manual(values = mtorf_colors, name = "mtORF Type") +
   theme(legend.position = "right",
        legend.margin = margin(0,0,0,0, unit = "mm"),
        plot.margin = margin(5,5,5,5, unit = "mm")) +
  ggtitle(tree_title)
dev.off()
```
```{r}
pdf("./outputs/figures/iqtree_nodes.pdf", width = 11.5, height = 8)
tree %>% ggtree() + geom_text(aes(label=node), size = 2, hjust=.45)
dev.off()
```


### RAxML
```{r}
tree <- read.newick("./outputs/raxml/RAxML_bestTree.best")
tree <- midpoint.root(tree)
gg_tree(tree)
```

