---
title: "Fst heatmaps"
author: "Mike Connelly"
date: "2024-02-09"
output: html_document
---

## Setup and packages
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, cache = TRUE)
options(stringsAsFactors = FALSE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```
```{r packages}
library("tidyverse")
#
library("pheatmap")
#
# library("qqman")
# library("fastman")
# library("OutFLANK")
#
library("patchwork")
#
source("./R/pocillopora_etp_gksim_functions.R")
```

## Import sample metadata
```{r import_sample_metadata}
source("./R/sample_import.R")
samples <- all_samples
```

## *Pocillopora* genomic spp.clusters
### Pairwise genome-wide Fst heatmaps
```{r}
# spp_combos <- data.frame(t(combn(unique(sort(samples$Species)), m = 2)))
#
spp_combos <- data.frame(t(combn(ngsadmix_spp, m = 2))) %>% 
  filter(!str_detect(`X2`, "Unassigned"))

cluster_combos <- data.frame(t(combn(ngsadmix_clusters, m = 2))) %>% 
  filter(!str_detect(`X2`, "Unassigned"))
```
```{r}
spp_fst <- read_delim("./outputs/angsd/pocillopora_spp_fst.txt",
                      delim = "\t",
                      col_names = c("spp1","spp2","Fst"),
                      col_types = cols("spp1" = col_factor(ordered = T),
                                     "spp2" = col_factor(ordered = T))) 
```

### Figure S11
```{r}
fst_heatmap <- ggplot(data = spp_fst, aes(spp1, spp2, fill = Fst)) +
  geom_tile(color = "white") +
  scale_fill_gradient(low = "white", high = "red", limits = c(0,0.75), name = expression(F[ST])) +
  geom_text(aes(label = round(Fst, 3)), size = 3) +
  theme(axis.title = element_blank(),
        axis.text.x = element_text(size = 10), 
        axis.text.y = element_text(size = 10),
        axis.ticks = element_blank(),
        panel.grid = element_blank(),
        panel.border = element_blank()) + 
  coord_fixed() +
  scale_x_discrete(position = "top")
print(fst_heatmap)
```
```{r}
pdf("./outputs/figures/FigS11_spp_fst_heatmap.pdf", width = 6.5, height = 3.75)
# title <- expression(italic("Pocillopora")*' species FST heatmap')
fst_heatmap 
dev.off()
```

### 50kb window Fst estimates
```{r}
angsd_fst_colnames <- c("SNP", "CHR", "BP", "Nsites", "Fst")
angsd_fst_cols <- cols("SNP" = col_character(),
                       "CHR" = col_character(),
                       "BP" = col_double(),
                       "Nsites" = col_double(),
                       "Fst" = col_double())
```

```{r}
spp_combos
```

```{r}
# for each pairwise comparison, read in 50kb window Fst and plot histogram and Fst manhattan plot
fst_files <- list.files("./outputs/angsd/50kb_fst/spp", full.names = T)
for (file in fst_files) { 
  fst_df <- read_delim(file, delim = "\t", skip = 1, col_names = angsd_fst_colnames, col_types = angsd_fst_cols)
  #
  fst_hist <- fst_df %>% ggplot(aes(Fst)) + 
    geom_histogram(binwidth = 0.02, color = "black", fill = "lightgrey") +
    xlim(c(0,1))
  #
  fst_plot <- fst_df %>%
    filter(!grepl("CA", CHR)) %>% 
    ggplot(aes(BP, Fst)) + 
    geom_point() +
    facet_grid(~CHR, scales = "free") +
    ylim(c(0,1)) +
    theme(axis.text.x = element_blank())
  #
  print((fst_plot / fst_hist) + plot_annotation(title = file))
}            
```

```{r fst_proportion_heatmap}
# plot a gradient of the proportion of windows in Fst bins
  fst_df_list <- list()
for (i in seq(1,10,1)) { 
  file = fst_files[i]
  fst_df <- read_delim(file, delim = "\t", skip = 1, col_names = angsd_fst_colnames, col_types = angsd_fst_cols)
  #
  fst_df_list[[i]] <- fst_df

fst_dist_df <- fst_df %>%
  # filter(Fst > 0) %>% 
  count(cut_width(Fst, 0.02)) %>% 
  rename("interval" = `cut_width(Fst, 0.02)`) %>% 
  mutate(prop = n/sum(n), comp = i)

fst_heatmap <- fst_dist_df %>%
  ggplot(aes(x = comp, y = interval, fill = prop))+
  geom_tile() +
  scale_fill_gradient(low="lightyellow", high="red")

print(fst_heatmap)
}
```
```{r}
# make clean plot of all spp. pairwise comparisons next to each other
fst_comparisons <- fst_files %>% 
  str_extract(., "/p.*_p.*_") %>% 
  str_remove(., "_50kb_reps_fst_") %>% # str_remove(., "_50kb_fst_") %>%
  str_remove(., "/") %>% 
  str_replace_all(., "peffusa", "Cluster 1") %>% 
  str_replace_all(., "pmeandrina", "Cluster 2") %>% 
  str_replace_all(., "pgrandisoffshore", "Cluster 3A") %>% 
  str_replace_all(., "pgrandiscontinent", "Cluster 3B") %>% 
  str_replace_all(., "pverrucosa3a", "Cluster 4") %>% 
  str_replace(., "_", " vs. ")

all_fst_df <- fst_df_list %>% 
  bind_rows(., .id = "column_label") %>% 
  mutate("comparison" = fst_comparisons[as.numeric(`column_label`)]) %>% 
  group_by(comparison) %>% 
  count(cut_width(Fst, 0.02)) %>% 
  rename("interval" = `cut_width(Fst, 0.02)`) %>% 
  group_by(comparison) %>% 
  mutate(prop = n/sum(n)) 

fst_50kb_dist_heatmap <- all_fst_df %>% 
  ggplot(aes(x = comparison, y = interval, fill = prop))+
  geom_tile() +
  scale_fill_gradient(low="white", high="red", name = "Proportion of \n50 kb windows") +
  ylab(expression(F[ST]*" "*intervals)) +
  #
geom_vline(xintercept = 1.5, color = "black", size = 0.25) +
geom_vline(xintercept = 2.5, color = "black", size = 0.25) +
geom_vline(xintercept = 3.5, color = "black", size = 0.25) +
geom_vline(xintercept = 4.5, color = "black", size = 0.25) +
geom_vline(xintercept = 5.5, color = "black", size = 0.25) +
geom_vline(xintercept = 6.5, color = "black", size = 0.25) +
geom_vline(xintercept = 7.5, color = "black", size = 0.25) +
geom_vline(xintercept = 8.5, color = "black", size = 0.25) +
geom_vline(xintercept = 9.5, color = "black", size = 0.25) +
  #
# geom_hline(yintercept = 3.5, size = 0.25) +
  theme(axis.text.y = element_blank(),
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 330, hjust = 0, size = 10),
        panel.grid = element_blank(),
        axis.ticks.y = element_blank())

print(fst_50kb_dist_heatmap)
```
### Figure S12
```{r}
pdf("./outputs/figures/FigS12_fst_50kb_distribution_heatmap.pdf", width = 6.5, height = 4)
print(fst_50kb_dist_heatmap)
dev.off()
```
### Plot Fst windows along chromosomes
```{r}
CHR_totlen <- fst_df_list[[i]] %>%
  filter(CHR %in% names(chrs_abbv)) %>% 
  # Compute chromosome size
  group_by(CHR) %>% 
  summarise(chr_len = max(BP) + 25000) %>% 
  # Calculate cumulative position of each chromosome
  mutate(tot = cumsum(chr_len) - chr_len) %>%
  select(-chr_len)
```
```{r}
pdf("./outputs/figures/fst_50kb_manhattan_plots.pdf", width = 6.5, height = 4.5)
for (i in seq(1,10,1)){
# join fst dataset to cumulative chromosome length
fst_df_tot  <- fst_df_list[[i]] %>% 
  filter(CHR %in% names(chrs_abbv)) %>% 
  left_join(CHR_totlen,  by = "CHR") %>% 
  # Add a cumulative position of each SNP
  arrange(CHR, BP) %>%
  mutate(BPtot = BP + tot)

# select top 0.1% of windows as "genomic islands of differentiation" (top 69 windows)
fst_outliers <- fst_df_tot %>% filter(CHR %in% names(chrs_abbv)) %>% slice_max(., prop = 0.01, order_by = Fst)

fst_manhattan_plot <- fst_df_tot %>% 
  ggplot(aes(BPtot,Fst)) +
  geom_hline(yintercept = mean(fst_df_tot$Fst)) +
  geom_point(aes(color = CHR), size = 0.5, show.legend = F) +
  geom_point(data = fst_outliers, aes(BPtot,Fst), color = "red", size = 0.75) +
  scale_color_manual(values = c(rep(c("grey25", "grey65"), 7))) +
  ylim(0,1)

print(fst_manhattan_plot)
}
dev.off()
```
```{r}
# select top 0.1% of windows as "genomic islands of differentiation" (top 69 windows)
fst_df_list[[5]] %>% filter(CHR %in% names(chrs_abbv)) %>% slice_max(., prop = 0.01, order_by = Fst)
```


## 10kb window Fst estimates
```{r}
fst_files <- list.files("./outputs/angsd/10kb_fst/spp", full.names = T)
for (file in fst_files) { 
  fst_df <- read_delim(file, delim = "\t", skip = 1, col_names = angsd_fst_colnames, col_types = angsd_fst_cols)
  #
  fst_hist <- fst_df %>% ggplot(aes(Fst)) + 
    geom_histogram(binwidth = 0.02, color = "black", fill = "lightgrey") +
    xlim(c(0,1))
  #
  fst_plot <- fst_df %>%
    filter(!grepl("CA", CHR)) %>% 
    ggplot(aes(BP, Fst)) + 
    geom_point() +
    facet_grid(~CHR, scales = "free") +
    ylim(c(0,1)) +
    theme(axis.text.x = element_blank())
  #
  print((fst_plot / fst_hist)) #+ plot_annotation(title = file))
}            
```


## *P. grandis* continental lineage populations
### Pairwise genome-wide Fst heatmaps
```{r}
# samples$Region <- factor(samples$Region, levels = rev(region_order_longitude), ordered = T)
pop_combos <- data.frame(t(combn(unique(sort(samples$Region)), m = 2))) %>% 
  filter(!str_detect(`X2`, "Clipperton"))
pop_combos
```
```{r}
pop_fst <- read_csv("./outputs/pgrandis_continental_population_fst.csv",
                    col_types = cols("pop1" = col_factor(ordered = T),
                                     "pop2" = col_factor(ordered = T)))
```


```{r}
fst_heatmap <- ggplot(data = pop_fst, aes(pop1, pop2, fill = Fst)) +
  geom_tile(color = "white") +
  scale_fill_gradient(low = "white", high = "red", limits = c(0,0.05), name = expression(F[ST])) +
  geom_text(aes(label = round(Fst, 3)), size = 2.5) +
  theme(axis.title = element_blank(),
        axis.text.x = element_text(angle = 30, vjust = 1, size = 8, hjust = 0), 
        axis.text.y = element_text(size = 8),
        axis.ticks = element_blank(),
        panel.grid = element_blank(),
        panel.border = element_blank()) + 
  coord_fixed() +
  scale_x_discrete(position = "top")
print(fst_heatmap)
```
```{r}
pdf("./outputs/figures/pgrandis_continental_fst_heatmap.pdf", height = 5, width = 5.75)
title <- expression(italic("P. grandis")*' population FST heatmap')
fst_heatmap + ggtitle(title)
dev.off()
```

```{r}
pop_fst_mat <- pop_fst %>% pivot_wider(names_from = "pop2", values_from = "Fst") %>% column_to_rownames("pop1") %>% as.matrix()
```
```{r}
# alternative approach with pheatmap
pheatmap(pop_fst_mat, cluster_cols = FALSE, cluster_rows = FALSE)
```

### 50kb window Fst estimates
```{r}
fst_files <- list.files("./outputs/angsd/10kb_fst/pgrandis_continental_pops", full.names = T)
for (file in fst_files) { 
  fst_df <- read_delim(file, delim = "\t", skip = 1, col_names = angsd_fst_colnames, col_types = angsd_fst_cols)
  #
  fst_hist <- fst_df %>% ggplot(aes(Fst)) + 
    geom_histogram(binwidth = 0.02, color = "black", fill = "lightgrey") +
    xlim(c(0,1))
  #
  fst_plot <- fst_df %>%
    filter(!grepl("CA", CHR)) %>% 
    ggplot(aes(BP, Fst)) + 
    geom_point(aes(color = CHR)) +
    # facet_grid(~CHR, scales = "free") +
    ylim(c(0,1)) +
    theme(axis.text.x = element_blank())
    
  #
  print((fst_plot / fst_hist) + plot_annotation(title = file))
}            
```
### 
```{r}
# pdf("./outputs/figures/10kbfst_pairwise.pdf")
fst_files <- list.files("./outputs/angsd/10kb_fst/spp_k6", full.names = T)
for (file in fst_files) { 
  fst_df <- read_delim(file, delim = "\t", skip = 1, col_names = angsd_fst_colnames, col_types = angsd_fst_cols)
  #
  fst_hist <- fst_df %>% ggplot(aes(Fst)) + 
    geom_histogram(binwidth = 0.02, color = "black", fill = "lightgrey") +
    xlim(c(0,1))
  #
  fst_plot <- fst_df %>% 
    filter(!grepl("CA", CHR)) %>% 
    ggplot(aes(BP, Fst)) + 
    geom_point() +
    facet_grid(~CHR, scales = "free") +
    ylim(c(0,1)) +
    theme(axis.text.x = element_blank())
  #
  print((fst_hist / fst_plot) + plot_annotation(title = file))
}
# dev.off()
```

## Experiment with "pretty" Manhattan plots
```{r}
fst_df %>% ggplot(aes(CHR, Fst)) + 
    geom_point() + 
    ylim(c(0,1)) 

summary(fst_df$Fst)
```
```{r}
fst_df %>% 
  filter(!grepl("CA", CHR)) %>% 
  ggplot(aes(BP, Fst)) + 
    geom_point() +
  facet_grid(~CHR, scales = "free") +
    ylim(c(0,1)) +
  theme(axis.text.x = element_blank())
``` 
