---
title: "Pipeline Metrics Summary"
author: "Mike Connelly"
date: "2023-08-07"
output: html_document
---
## Setup and packages
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, cache = TRUE)
options(stringsAsFactors = FALSE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```
```{r packages}
library("tidyverse")
#
source("./R/pocillopora_etp_gksim_functions.R")
```

## Colors and theme
```{r colors}
region_colors <- c("gold1", "steelblue", "darkorange", "darkturquoise")
#
spp_colors <- c("purple", "darkorange1", "gold", "turquoise2", "turquoise4")
# 
theme_set(theme_bw())
```

## Import sample metadata
```{r import_sample_metadata}
all_samples <- read_csv("./data/pocillopora_samples_all_metadata.csv")
#
# select_samples <- read_csv("./data/etp_samples.txt")
select_samples <- read_csv("./data/phyluce_noclones_samples.txt", col_names = c("Sample ID"))
```
```{r select_samples}
samples <- filter(all_samples, all_samples$`Sample ID` %in% select_samples$`Sample ID`)
```
## Sample summaries
```{r}
## Panama regions and species
samples %>% filter(Region == "Coiba" | Region == "Perlas") %>% group_by(Region, Species) %>% count()

## Panama P. grandis site and regions
samples %>% filter(Region == "Coiba" | Region == "Perlas") %>% filter(Species == "P. grandis") %>% group_by(`Region:Site`) %>% count()
```

## Pipeline metrics summaries
### Raw and trimmed reads
```{r import_multiqc_data}
# raw reads
multiqc_fastqc <- read_tsv("./outputs/QCs/fastqcs/multiqc_data/multiqc_fastqc.txt") %>%
  mutate("Sample ID" = gsub("_S.*_R[12]_001", "", Sample)) %>%
  mutate(`Sample ID` = gsub("^05_", "", `Sample ID`))

# trimmed reads
multiqc_trimqc <- read_tsv("./outputs/QCs/trimqcs/multiqc_data/multiqc_fastqc.txt") %>%
  mutate("Sample ID" = gsub("_S.*_R[12]_PE", "", Sample))
```

```{r tidy_multiqc_data}
# tidy multiqc data to look at raw versus trimmed reads
multiqc_fastqc_reads <- multiqc_fastqc %>% filter(grepl("R1", Sample)) %>% select(`Sample ID`, `Total Sequences`) %>% rename("Raw Reads" = `Total Sequences`)

multiqc_trimqc_reads <- multiqc_trimqc %>% filter(grepl("R1", Sample)) %>% select(`Sample ID`, `Total Sequences`) %>% rename("Trimmed Reads" = `Total Sequences`)
#
multiqc_data <- left_join(samples, multiqc_fastqc_reads, by = "Sample ID") %>% left_join(., multiqc_trimqc_reads, by = "Sample ID")
```

```{r multiqc_plot_data}
multiqc_plot_data <- multiqc_data %>% pivot_longer(., `Raw Reads`:`Trimmed Reads`, names_to = "pipeline_step", values_to = "read_count") %>%
  dplyr::group_by(`Sample ID`) %>%
  dplyr::mutate(proportion = read_count / read_count[pipeline_step == "Raw Reads"])
# 
multiqc_plot_data$pipeline_step <- factor(multiqc_plot_data$pipeline_step, levels = c("Raw Reads", "Trimmed Reads"), ordered = TRUE) # , "Uniquely Mapped Reads", "Input Reads", "Assigned Reads"
```

```{r read_depth_plot}
# All sequenced samples
FASTQCg <- multiqc_plot_data %>% 
  filter(pipeline_step == "Raw Reads") %>%
  ggplot(aes(`Sample ID`, read_count))
FASTQCg + geom_point() + 
  geom_hline(yintercept = 10000000, color  = "green") +
  geom_hline(yintercept = 8000000) + 
  geom_hline(yintercept = 1000000, color = "red")
```

### BWA mapped reads

```{r}

```

## Pipeline metrics plot
```{r}
pipeQC_counts <- multiqc_plot_data %>%
  # dplyr::filter(pipeline_step != "Input Reads" & pipeline_step != "Assigned Reads") %>%
  ggplot(aes(pipeline_step, read_count)) +
  geom_path(aes(group = `Sample ID`), show.legend = FALSE) +
  geom_point(aes(fill = Species), size = 2, shape = 21, color = "black", alpha = 0.5, stroke = 0.5, show.legend = FALSE) + 
  facet_grid(.~Region, scales = "free") +
  scale_y_continuous(limits = c(0, 3e7), breaks = c(0, 1e7, 2e7, 3e7)) +
  # scale_color_manual(values = genocolors) +
  # scale_shape_manual(values = colshapes) + 
  # scale_fill_manual(values = genocolors) +
  theme(axis.text.x = element_text(angle = 315, hjust = 0),
        axis.title.x = element_blank()
        # panel.grid.major = element_line(color = "grey92"),
        ) +
  ylab("read count") +
  xlab("pipeline step") + 
  ggtitle("read counts")
pipeQC_counts
```
```{r}
pipeQC_proportions <- multiqc_plot_data %>%
  # dplyr::filter(pipeline_step != "Input Reads" & pipeline_step != "Assigned Reads") %>%
  ggplot(aes(pipeline_step, proportion)) +
  geom_path(aes(group = `Sample ID`), show.legend = FALSE) +
  geom_point(aes(fill = Species), size = 2, shape = 21, color = "black", alpha = 0.5, stroke = 0.5, show.legend = FALSE) + 
  # facet_grid(.~Region, scales = "free") +
  scale_y_continuous(limits = c(0, 1)) +
  # scale_color_manual(values = genocolors) +
  # scale_shape_manual(values = colshapes) + 
  # scale_fill_manual(values = genocolors) +
  theme(axis.text.x = element_text(angle = 315, hjust = 0),
        axis.title.x = element_blank()
        # panel.grid.major = element_line(color = "grey92"),
        ) +
  ylab("proportion raw reads") +
  xlab("pipeline step") + 
  ggtitle("read proportions")
pipeQC_proportions
```


SPAdes assembly metrics

PHYLUCE alignment matrix stats

SNP matrix stats